---
title: "Benchmarks"
author:
  - name: "Emir Turkes"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(
    inputFile, encoding = encoding,
    output_file = file.path(
      "..", "results", "benchmarks.html"
    ))})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 25px; text-indent: 0;
}
.tocify-subheader .tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 35px; text-indent: 0;
}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of [genefunnel-benchmarks](https://github.com/eturkes/genefunnel-benchmarks).*

The table of contents in the top left is clickable and can be used to quickly navigate the document.
To toggle the visibility of code, use the `CODE` toggles at the top right of chunks.
The toggle at the start of the document controls the visibility of all chunks.

```{r}
#    This file is part of genefunnel-benchmarks.
#    Copyright (C) 2024  Emir Turkes, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

# Load required packages, suppressing startup messages.
# -----------------------------------------------------
library(conflicted)
conflicts_prefer(base::intersect, edgeR::cpm, .quiet = TRUE)
packages <- c(
  "GSVA", "Rcpp", "Matrix", "ComplexHeatmap", "circlize", "Seurat", "GSEABase",
  "edgeR", "SingleCellExperiment", "scuttle"
)
invisible(
  suppressPackageStartupMessages(
    lapply(packages, FUN = library, character.only = TRUE)
  )
)

sourceCpp(file.path("..", "src", "calculateScores.cpp"))
source("utils.R")

`%notin%` <- Negate(`%in%`)
# -----------------------------------------------------

# Define global settings.
# -----------------------
knitr::opts_chunk$set(fig.width = 7, fig.height = 7, dpi = 72)
ht_opt$DIMNAME_PADDING = unit(3, units = "mm")
# -----------------------

# Useful variables.
# -----------------
data_dir <- file.path("..", "data")
cache_dir <- file.path("..", "cache")
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}
# -----------------
```

# Simulated Scenarios

```{r}
vec <- c(
  rep(50, times = 10),
  rep(95, times = 5), rep(5, times = 5),
  rep(95, times = 10),
  rep(5, times = 10),
  c(c(9995, 9995, rep(5, times = 3)), c(9995, rep(5, times = 4)))
)
set.seed(1)
noise <- round(runif(50, min = -4, max = 4))

mat <- matrix(
  vec + noise,
  nrow =  10,
  ncol = 5,
  dimnames = list(
    paste("Gene", toupper(letters[1:10])),
    c("Sample 1", "Sample 2", "Sample 3", "Sample 4", "Sample5")
  )
)
# mat <- log2(mat)

top_anno <- HeatmapAnnotation(
  sample = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = c("Sample 1", "Sample 2", "Sample 3", "Sample 4", "Sample5"),
    labels_gp = gpar(col = "black", fontsize = 15),
    height = unit(10, "mm")
  )
)
row_anno <- rowAnnotation(
  gene_sets = anno_block(labels = c("Gene Set X", "Gene Set Y")),
  empty = anno_empty(border = FALSE, width = unit(8, units = "mm"))
)

col_range <- c(min(mat), (min(mat) + max(mat)) / 2, max(mat))
draw(
  Heatmap(
    mat,
    name = "Gene Expression",
    col = colorRamp2(col_range, colors = c("blue", "white", "red")),
    cluster_rows = FALSE,
    cluster_columns = FALSE,
    show_column_names = FALSE,
    row_title = NULL,
    top_annotation = top_anno,
    right_annotation = row_anno,
    column_split = seq(5),
    row_split = c(rep(1, times = 5), rep(2, times = 5)),
    column_title = "Hypothetical Raw Gene Count by Sample Matrix",
    rect_gp = gpar(col = "black", lwd = 0.5),
    column_gap = unit(1, units = "mm"),
    row_names_gp = gpar(fontsize = 15),
    column_title_gp = gpar(fontsize = 18),
    row_names_side = "left",
    heatmap_legend_param = list(
      labels_gp = gpar(fontsize = 14), title_gp = gpar(fontsize = 15),
      border = "lightgrey", title_position = "topcenter",
      grid_height = unit(4, "mm"), direction = "horizontal"
    )
  ),
  heatmap_legend_side = "top"
)

seekViewport("annotation_empty_1")
loc1 = deviceLoc(x = unit(0, "npc"), y = unit(0, "npc"))
seekViewport("annotation_empty_2")
loc2 = deviceLoc(x = unit(1, "npc"), y = unit(1, "npc"))

seekViewport("global")
grid.rect(
  loc2$x * 0.96, (loc1$y + loc2$y) / 2,
  width = loc2$x - loc1$x, height = loc2$y * 1.96,
  just = c("left", "centre")
)
grid.text(
  "Gene Set Z", rot = 90,
  x = (loc1$x + loc2$x) * 0.502, y = (loc1$y + loc2$y) * 0.5
)

log_mat <- log1p(mat)
col_range <- c(min(log_mat), (min(log_mat) + max(log_mat)) / 2, max(log_mat))
draw(
  Heatmap(
    log_mat,
    name = "log1p Gene Expression",
    col = colorRamp2(col_range, colors = c("blue", "white", "red")),
    cluster_rows = FALSE,
    cluster_columns = FALSE,
    show_column_names = FALSE,
    row_title = NULL,
    top_annotation = top_anno,
    right_annotation = row_anno,
    column_split = seq(5),
    row_split = c(rep(1, times = 5), rep(2, times = 5)),
    column_title = "Hypothetical Raw Gene Count by Sample Matrix",
    rect_gp = gpar(col = "black", lwd = 0.5),
    column_gap = unit(1, units = "mm"),
    row_names_gp = gpar(fontsize = 15),
    column_title_gp = gpar(fontsize = 18),
    row_names_side = "left",
    heatmap_legend_param = list(
      labels_gp = gpar(fontsize = 14), title_gp = gpar(fontsize = 15),
      border = "lightgrey", title_position = "topcenter",
      grid_height = unit(4, "mm"), direction = "horizontal"
    )
  ),
  heatmap_legend_side = "top"
)

seekViewport("annotation_empty_1")
loc1 = deviceLoc(x = unit(0, "npc"), y = unit(0, "npc"))
seekViewport("annotation_empty_2")
loc2 = deviceLoc(x = unit(1, "npc"), y = unit(1, "npc"))

seekViewport("global")
grid.rect(
  loc2$x * 0.96, (loc1$y + loc2$y) / 2,
  width = loc2$x - loc1$x, height = loc2$y * 1.96,
  just = c("left", "centre")
)
grid.text(
  "Gene Set Z", rot = 90,
  x = (loc1$x + loc2$x) * 0.502, y = (loc1$y + loc2$y) * 0.5
)
```

```{r, results = "hide"}
gene_sets <- list(
  `Gene Set X` = rownames(mat)[1:5], `Gene Set Y` = rownames(mat)[6:10],
  `Gene Set Z` = rownames(mat)
)

out <- vector("list", length = 5)
names(out) <- c("GSVA", "ssGSEA", "PLAGE", "Z-score", "GeneFunnel")

tmp <- gsvaParam(mat, gene_sets, kcdf = "Poisson")
out[[1]] <- gsva(tmp, verbose = FALSE)

tmp <- ssgseaParam(log2(mat), gene_sets, normalize = FALSE)
out[[2]] <- gsva(tmp, verbose = FALSE)

tmp <- zscoreParam(log2(mat), gene_sets)
out[[3]] <- gsva(tmp, verbose = FALSE)

tmp <- plageParam(log2(mat), gene_sets)
out[[4]] <- gsva(tmp, verbose = FALSE)

out[[5]] <- genefunnel(mat, gene_sets)
```

```{r, fig.width = 5, fig.height = 3}
top_anno <- HeatmapAnnotation(
  sample = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = c("S1", "S2", "S3", "S4", "S5"),
    labels_gp = gpar(col = "black", fontsize = 15),
    height = unit(10, "mm")
  )
)

for (i in seq_along(out)) {
  col_range <- c(
    min(out[[i]]), (min(out[[i]]) + max(out[[i]])) / 2, max(out[[i]])
  )
  if (!all(is.na(col_range))) {
    ht <- draw(
      Heatmap(
        out[[i]],
        name = "Gene Set Score",
        col = colorRamp2(col_range, colors = c("blue", "white", "red")),
        cluster_rows = FALSE,
        cluster_columns = FALSE,
        show_column_names = FALSE,
        top_annotation = top_anno,
        column_split = seq(5),
        column_title = names(out)[[i]],
        rect_gp = gpar(col = "black", lwd = 0.5),
        column_gap = unit(1, units = "mm"),
        row_names_gp = gpar(fontsize = 16),
        column_title_gp = gpar(fontsize = 18),
        row_names_side = "left",
        heatmap_legend_param = list(
          labels_gp = gpar(fontsize = 14), title_gp = gpar(fontsize = 15),
          border = "lightgrey", title_position = "topcenter",
          grid_height = unit(4, "mm"), direction = "horizontal"
        )
      ),
      heatmap_legend_side = "top",
    )
    ht
  } else {
    ht <- draw(
      Heatmap(
        out[[i]],
        name = "Gene Set Score",
        cluster_rows = FALSE,
        cluster_columns = FALSE,
        show_column_names = FALSE,
        top_annotation = top_anno,
        column_split = seq(5),
        column_title = names(out)[[i]],
        rect_gp = gpar(col = "black", lwd = 0.5),
        column_gap = unit(1, units = "mm"),
        row_names_gp = gpar(fontsize = 16),
        column_title_gp = gpar(fontsize = 18),
        row_names_side = "left",
        heatmap_legend_param = list(
          labels_gp = gpar(fontsize = 14), title_gp = gpar(fontsize = 15),
          border = "lightgrey", title_position = "topcenter",
          grid_height = unit(4, "mm"), direction = "horizontal"
        )
      ),
      heatmap_legend_side = "top",
    )
    ht
  }
}
```

# Simulated scRNAseq dataset

```{r, results = "hide"}
rds <- file.path(cache_dir, "gene_anno.rds")
rds2 <- file.path(cache_dir, "sim.rds")
if (file.exists(rds)) {
  gene_anno <- readRDS(rds)
  sim <- readRDS(rds2)
} else {
  seurat <- readRDS(file.path(data_dir, "annotated_seurat.rds"))
  DefaultAssay(seurat) <- "RNA"
  seurat@active.ident <- seurat$ctype
  seurat <- subset(
    seurat,
    idents = c("EX L2-4 CUX2", "EX L4-5 RORB", "EX L5 RORB", "EX L5-6 THEMIS")
  )
  seurat@active.ident <- seurat$tau
  seurat <- subset(seurat, idents = "CTRL")
  gc()

  sim1 <- SingleCellExperiment(
    list(counts = GetAssayData(seurat, layer = "counts"))
  )
  sim2 <- sim1
  sim3 <- sim1
  sim <- cbind(sim1, sim2, sim3, deparse.level = 1)

  ncol <- ncol(seurat)
  sim$Group <- factor(
    c(
      rep("Group1", times = ncol),
      rep("Group2", times = ncol),
      rep("Group3", times = ncol)
    )
  )

  gene_anno <- data.frame(
    symbol = unlist(seurat@misc), ensembl = rownames(seurat)
  )

  rm(sim1, sim2, sim3, seurat)
  gc()
  saveRDS(gene_anno, file = rds)
  saveRDS(sim, file = rds2)
}

gene_sets <- readRDS(file.path(data_dir, "gene_sets.rds"))

sim <- suppressWarnings(
  aggregateAcrossCells(
    sim, colData(sim)[ , "Group"],
    use_exprs_values = "counts", statistics = "sum"
  )
)

mat <- counts(sim)
mean <- mean(mat[ , 1])
mat[ , 2] <- rep(mean, times = nrow(mat))
mat[ , 3] <- rep(c(mean / 4, mean * 4), times = nrow(mat) / 2)
counts(sim) <- mat

rds <- file.path(cache_dir, "out.rds")
if (file.exists(rds)) {
  out <- readRDS(rds)
} else {

  out <- vector("list", length = 5)
  names(out) <- c("GSVA", "ssGSEA", "PLAGE", "Z-score", "GeneFunnel")

  # tmp <- gsvaParam(counts(sim), gene_sets, kcdf = "Poisson")
  # out[[1]] <- gsva(tmp, verbose = FALSE)
  # 
  # tmp <- DGEList(counts(sim))
  # tmp$samples$group <- sim$Group
  # keep <- which(rowSums(tmp$counts) > 0)
  # tmp <- tmp[keep, , keep.lib.sizes = FALSE]
  # tmp <- calcNormFactors(tmp)
  # tmp <- voomLmFit(tmp)
  # tmp <- ssgseaParam(tmp$EList$E, gene_sets, normalize = FALSE)
  # out[[2]] <- gsva(tmp, verbose = FALSE)

  # tmp <- zscoreParam(logcounts(sim), gene_sets)
  # out[[3]] <- gsva(tmp, verbose = FALSE)

  # tmp <- plageParam(logcounts(sim), gene_sets)
  # out[[4]] <- gsva(tmp, verbose = FALSE)

  out[[5]] <- genefunnel(counts(sim), geneIds(gene_sets))

  saveRDS(out, file = rds)
}
```

```{r, fig.width = 9, fig.height = 5}
top_anno <- HeatmapAnnotation(
  sample = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = c("Group1", "Group2", "Group3"),
    labels_gp = gpar(col = "black", fontsize = 12),
    height = unit(8, "mm")
  )
)

for (i in seq_along(out)) {
  if (!is.null(out[[i]])) {
    mat <- SingleCellExperiment(list(counts = out[[i]]))
    mat$Group <- sim$Group
    col_range <- c(
      min(counts(mat)),
      (min(counts(mat)) + max(counts(mat))) / 2,
      max(counts(mat))
    )

    tmp <- mat[order(counts(mat[ , 1]), decreasing = TRUE), ]
    rownames <- rownames(tmp)[1:5]
    tmp <- mat[order(counts(mat[ , 2]), decreasing = TRUE), ]
    rownames <- c(rownames, rownames(tmp)[1:5])
    tmp <- mat[order(counts(mat[ , 3]), decreasing = TRUE), ]
    rownames <- c(rownames, rownames(tmp)[1:5])
    mat <- mat[rownames(mat) %in% rownames, ]
    mat <- mat[match(rownames, rownames(mat)), ]
    ht <- draw(
      Heatmap(
        counts(mat),
        name = "Gene Set Score",
        col = colorRamp2(col_range, colors = c("blue", "white", "red")),
        cluster_rows = FALSE,
        cluster_columns = FALSE,
        show_column_names = FALSE,
        top_annotation = top_anno,
        column_split = mat$Group,
        column_title = names(out)[[i]],
        rect_gp = gpar(col = "black", lwd = 0.5),
        column_gap = unit(1, units = "mm"),
        row_names_gp = gpar(fontsize = 12),
        column_title_gp = gpar(fontsize = 16),
        row_names_side = "left",
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 12)
        ),
        heatmap_legend_param = list(
          labels_gp = gpar(fontsize = 10), title_gp = gpar(fontsize = 12),
          border = "lightgrey", title_position = "topcenter",
          grid_height = unit(4, "mm"), direction = "horizontal"
        )
      ),
      heatmap_legend_side = "top",
    )
    ht
  }
}
```

```{r, fig.width = 6, fig.height = 7}
ht_opt$DIMNAME_PADDING = unit(2, units = "mm")

col_range <- c(
  min(counts(sim)),
  (min(counts(sim)) + max(counts(sim))) / 2,
  max(counts(sim))
)

gse_plot <- rownames(mat)
for (gse in gse_plot) {
  genes <- unlist(geneIds(gene_sets[names(gene_sets) %in% gse]))
  mat <- sim[rownames(sim) %in% genes, ]
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(mat), ]
  gene_anno_sub <- gene_anno_sub[match(rownames(mat), gene_anno_sub$ensembl), ]
  rownames(mat) <- gene_anno_sub$symbol
  ht <- draw(
    Heatmap(
      counts(mat),
      name = "Raw Gene Expression",
      col = colorRamp2(col_range, colors = c("blue", "white", "red")),
      cluster_rows = FALSE,
      cluster_columns = FALSE,
      show_column_names = FALSE,
      top_annotation = top_anno,
      column_split = mat$Group,
      column_title = gse,
      rect_gp = gpar(col = "black", lwd = 0.5),
      column_gap = unit(1, units = "mm"),
      row_names_gp = gpar(fontsize = 9),
      column_title_gp = gpar(fontsize = 12),
      row_names_side = "left",
      heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 10), title_gp = gpar(fontsize = 12),
        border = "lightgrey", title_position = "topcenter",
        grid_height = unit(4, "mm"), direction = "horizontal"
      )
    ),
    heatmap_legend_side = "top",
  )
  ht
}

logcounts(sim) <- log1p(counts(sim))
col_range <- c(
  min(logcounts(sim)),
  (min(logcounts(sim)) + max(logcounts(sim))) / 2,
  max(logcounts(sim))
)

for (gse in gse_plot) {
  genes <- unlist(geneIds(gene_sets[names(gene_sets) %in% gse]))
  mat <- sim[rownames(sim) %in% genes, ]
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(mat), ]
  gene_anno_sub <- gene_anno_sub[match(rownames(mat), gene_anno_sub$ensembl), ]
  rownames(mat) <- gene_anno_sub$symbol
  ht <- draw(
    Heatmap(
      logcounts(mat),
      name = "log1p Gene Expression",
      col = colorRamp2(col_range, colors = c("blue", "white", "red")),
      cluster_rows = FALSE,
      cluster_columns = FALSE,
      show_column_names = FALSE,
      top_annotation = top_anno,
      column_split = mat$Group,
      column_title = gse,
      rect_gp = gpar(col = "black", lwd = 0.5),
      column_gap = unit(1, units = "mm"),
      row_names_gp = gpar(fontsize = 9),
      column_title_gp = gpar(fontsize = 12),
      row_names_side = "left",
      heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 10), title_gp = gpar(fontsize = 12),
        border = "lightgrey", title_position = "topcenter",
        grid_height = unit(4, "mm"), direction = "horizontal"
      )
    ),
    heatmap_legend_side = "top",
  )
  ht
}
```

# References

This is the concluding section of the document, where we output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
