---
title: "GeneFunnel Benchmarks Viewer"
output:
  flexdashboard::flex_dashboard:
    orientation: rows
    runtime: shiny
---

```{r global, include = FALSE}
#    This file is part of genefunnel-benchmarks.
#    Copyright (C) 2024-2025  Emir Turkes, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

library(conflicted)
packages <- c(
  "Rcpp", "Matrix", "shinyMatrix", "ComplexHeatmap", "circlize", "GSVA",
  "BiocParallel", "parallelly", "cowplot", "knitr", "kableExtra", "GSEABase",
  "scuttle", "Seurat", "edgeR", "bench", "ggplot2"
)
lapply(packages, FUN = library, character.only = TRUE)

`%notin%` <- Negate(`%in%`)

sourceCpp(file.path("..", "src", "calculateScores.cpp"))
source("utils.R")

data_dir <- file.path("..", "data")
cache_dir <- file.path("..", "cache")

ht_opt$DIMNAME_PADDING <- unit(3, units = "mm")

top_anno <- HeatmapAnnotation(
  sample = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = c("Sample 1", "Sample 2", "Sample 3", "Sample 4", "Sample 5"),
    labels_gp = gpar(col = "black", fontsize = 15),
    height = unit(10, "mm")
  )
)
top_anno2 <- HeatmapAnnotation(
  sample = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = c("Sample 1", "Sample 2", "Sample 3", "Sample 4"),
    labels_gp = gpar(col = "black", fontsize = 15),
    height = unit(10, "mm")
  )
)
top_anno3 <- HeatmapAnnotation(
  sample = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = c("S1", "S2", "S3", "S4"),
    labels_gp = gpar(col = "black", fontsize = 15),
    height = unit(10, "mm")
  )
)
top_anno4 <- HeatmapAnnotation(
  sample = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = c("Original", "Modified"),
    labels_gp = gpar(col = "black", fontsize = 12),
    height = unit(8, "mm")
  )
)
top_anno5 <- HeatmapAnnotation(
  sample = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = c("Original", "Low Var.", "High Var."),
    labels_gp = gpar(col = "black", fontsize = 10),
    height = unit(8, "mm")
  )
)
top_anno6 <- HeatmapAnnotation(
  sample = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = c("NFT", "CTRL"),
    labels_gp = gpar(col = "black", fontsize = 12),
    height = unit(8, "mm")
  )
)
top_anno7 <- HeatmapAnnotation(
  sample = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = c("Sample 1", "Sample 2", "Sample 3", "Sample 4", "Sample 5"),
    labels_gp = gpar(col = "black", fontsize = 12),
    height = unit(9, "mm")
  )
)
top_anno8 <- HeatmapAnnotation(
  sample = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = c("Original", "Modified"),
    labels_gp = gpar(col = "black", fontsize = 7),
    height = unit(5, "mm")
  )
)
row_anno <- rowAnnotation(
  gene_sets = anno_block(labels = c("Gene Set X", "Gene Set Y")),
  empty = anno_empty(border = FALSE, width = unit(8, units = "mm"))
)

colnames <- c(
  "------Sample 1",
  "------Sample 2",
  "------Sample 3",
  "------Sample 4"
)

vec <- c(
  rep(50, times = 10),
  c(90, 10, 90, 10, 90, 90, 10, 10, 90, 10),
  rep(100, times = 10),
  rep(100, times = 5), rep(NA, times = 5),
  rep(100, times = 1), rep(0, times = 9)
)
set.seed(4)
noise1 <- round(runif(10, min = -5, max = 5))
noise1[1] <- 2
noise2 <- noise1 * 2
noise2[1] <- -90
noise2[3] <- noise2[3] + 42
noise2[5] <- noise2[5] + 42
noise2[7] <- noise2[7] + 10
noise2 <- noise2[c(1:5, 7, 6, 8:10)]
noise3 <- noise1 * 2
noise4 <- rep(0, times = 10)
noise5 <- noise4

default_mat <- matrix(
  vec + c(noise1, noise2, noise5, noise4, noise5),
  nrow =  10,
  ncol = 5,
  dimnames = list(
    paste("Gene", toupper(letters[1:10])),
    c("Sample 1", "Sample 2", "Sample 3", "Sample 4", "Sample 5")
  )
)
gene_sets <- list(
  `Gene Set X` = rownames(default_mat)[1:5],
  `Gene Set Y` = rownames(default_mat)[6:10],
  `Gene Set Z` = rownames(default_mat)
)

rds <- file.path(cache_dir, "gene_anno.rds")
rds2 <- file.path(cache_dir, "sim.rds")
if (file.exists(rds)) {
  gene_anno <- readRDS(rds)
  sim <- readRDS(rds2)
} else {
  seurat <- readRDS(file.path(data_dir, "annotated_seurat.rds"))
  DefaultAssay(seurat) <- "RNA"
  seurat@active.ident <- seurat$ctype
  seurat <- subset(
    seurat,
    idents = c("EX L2-4 CUX2", "EX L4-5 RORB", "EX L5 RORB", "EX L5-6 THEMIS")
  )
  seurat@active.ident <- seurat$tau
  seurat_CTRL <- seurat
  seurat <- subset(seurat, idents = "NFT")
  gc()

  sim1 <- SingleCellExperiment(
    list(counts = GetAssayData(seurat, layer = "counts"))
  )
  sim2 <- sim1
  sim3 <- sim1
  sim4 <- sim1
  sim5 <- sim1
  sim6 <- sim1
  sim7 <- sim1

  seurat_CTRL <- subset(seurat_CTRL, idents = "CTRL")
  sim8 <- SingleCellExperiment(
    list(counts = GetAssayData(seurat_CTRL, layer = "counts"))
  )

  sim <- cbind(sim1, sim2, sim3, sim4, sim5, sim6, sim7, sim8, deparse.level = 1)

  ncol <- ncol(seurat)
  sim$Group <- factor(
    c(
      rep("Group 1", times = ncol),
      rep("Group 2", times = ncol),
      rep("Group 3", times = ncol),
      rep("Group 4", times = ncol),
      rep("Group 5", times = ncol),
      rep("Group 6", times = ncol),
      rep("Group 7", times = ncol),
      rep("Group 8", times = ncol(seurat_CTRL))
    )
  )

  gene_anno <- data.frame(
    symbol = unlist(seurat@misc), ensembl = rownames(seurat)
  )

  rm(sim1, sim2, sim3, sim4, sim5, sim6, sim7, sim8, seurat, seurat_CTRL)
  gc()
  saveRDS(gene_anno, file = rds)
  saveRDS(sim, file = rds2)
}

gene_sets_sim <- readRDS(file.path(data_dir, "gene_sets_sim.rds"))

sim <- sim[rownames(sim) %in% unlist(geneIds(gene_sets_sim)), ]
sc <- sim
sim <- suppressWarnings(
  aggregateAcrossCells(
    sim, colData(sim)[ , "Group"],
    use_exprs_values = "counts", statistics = "sum",
    BPPARAM = MulticoreParam(availableCores())
  )
)

adjust_variability <- function(orig_vector, high_var = FALSE) {

  orig_sum <- sum(orig_vector)
  orig_mean <- mean(orig_vector)
  scaled_values <- orig_vector

  if (high_var) {
    scaling_factors <- 1 + (0.01 * abs(scaled_values - orig_mean) / orig_mean)
  } else {
    scaling_factors <- 1 * abs(scaled_values - orig_mean) /
      max(abs(scaled_values - orig_mean))
    replace <- which(scaling_factors >= 1)
    scaling_factors <- 1 - scaling_factors
  }
  scaled_values <- orig_mean + ((scaled_values - orig_mean) * scaling_factors)

  if (high_var) {
    scaled_values[scaled_values < 0] <- 0
    scaled_values <- (scaled_values / sum(scaled_values)) * orig_sum
  } else {
    for (idx in replace) {
      scaled_values[idx] <- orig_vector[idx]
    }
    scaled_values <- scaled_values * (orig_sum / sum(scaled_values))
  }

  integer_values <- round(scaled_values)
  diff <- orig_sum - sum(integer_values)
  adjust_idx <- order(abs(scaled_values - integer_values), decreasing = TRUE)
  for (i in seq_len(abs(diff))) {
    if (diff > 0) {
      integer_values[adjust_idx[i]] <- integer_values[adjust_idx[i]] + 1
    } else if (diff < 0) {
      integer_values[adjust_idx[i]] <- integer_values[adjust_idx[i]] - 1
    }
  }

  return(integer_values)
}

mat <- counts(sim)
mat[ , 2] <- adjust_variability(mat[ , 1], high_var = FALSE)
mat[ , 3] <- adjust_variability(mat[ , 1], high_var = TRUE)

genes <- unlist(
  geneIds(
    gene_sets_sim[names(gene_sets_sim) == "NELF complex"]
  )
)
mat[which(rownames(mat) %in% genes), 4] <- round(
  sum(mat[which(rownames(mat) %in% genes), 4]) / length(genes)
)

genes <- unlist(
  geneIds(
    gene_sets_sim[names(gene_sets_sim) == "trace-amine receptor activity"]
  )
)
mat[which(rownames(mat) %in% genes), 4] <-
  mat[which(rownames(mat) %in% genes), 4] + 1000

genes <- unlist(
  geneIds(
    gene_sets_sim[names(gene_sets_sim) == "tau protein binding"]
  )
)
mat[which(rownames(mat) %in% genes), 5] <- round(
  sum(mat[which(rownames(mat) %in% genes), 5]) / length(genes)
)

genes <- unlist(
  geneIds(
    gene_sets_sim[names(gene_sets_sim) == "neurofibrillary tangle"]
  )
)
mat[which(rownames(mat) %in% genes), 6] <-
  mat[which(rownames(mat) %in% genes), 6] + 1000

dge_orig <- DGEList(mat[ , c(7, 8)])
keep <- filterByExpr(
  dge_orig, group = sim$Group,
  min.count = 10, min.total.count = 15
)
dge_orig <- dge_orig[keep, ]
dge_orig <- calcNormFactors(dge_orig)
dge <- DGEList(mat[ , c(7, 8)])
dge$samples$norm.factors <- dge_orig$samples$norm.factors
dge <- calcNormFactors(dge)
mat[ , c(7, 8)] <- edgeR::cpm(
  dge, log = FALSE, prior.count = 0
)

counts(sim) <- mat
logcounts(sim) <- log2(mat + 1)

rds <- file.path(cache_dir, "out_sim.rds")
if (file.exists(rds)) {
  out_sim <- readRDS(rds)
} else {

  out_sim <- vector("list", length = 6)
  names(out_sim) <- c(
    "GSVA Poisson", "GSVA Gaussian", "ssGSEA", "PLAGE", "Z-score", "GeneFunnel"
  )

  tmp <- gsvaParam(
    counts(sim), gene_sets_sim, minSize = 2, kcdf = "Poisson"
  )
  out_sim[[1]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

  tmp <- gsvaParam(
    logcounts(sim), gene_sets_sim, minSize = 2, kcdf = "Gaussian"
  )
  out_sim[[2]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

  tmp <- ssgseaParam(
    logcounts(sim), gene_sets_sim, minSize = 2, normalize = FALSE
  )
  out_sim[[3]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

  tmp <- zscoreParam(
    logcounts(sim), gene_sets_sim, minSize = 2
  )
  out_sim[[4]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

  tmp <- plageParam(
    logcounts(sim), gene_sets_sim, minSize = 2
  )
  out_sim[[5]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

  out_sim[[6]] <- genefunnel(
    counts(sim), geneIds(gene_sets_sim),
    BPPARAM = MulticoreParam(availableCores())
  )
  out_sim[[6]] <- out_sim[[6]]

  saveRDS(out_sim, file = rds)
}

sim_sub <- sim[ , 1:6]
counts(sim_sub) <- Matrix(counts(sim_sub), sparse = TRUE)
logcounts(sim_sub) <- Matrix(logcounts(sim_sub), sparse = TRUE)

rds <- file.path(cache_dir, "bench.rds")
if (file.exists(rds)) {
  bench <- readRDS(rds)
} else {

  gc()
  bench <- mark(
    gsva(
      gsvaParam(
        counts(sim_sub), gene_sets_sim, minSize = 2, kcdf = "Poisson"
      ),
      BPPARAM = SerialParam()
    ),
    gsva(
      gsvaParam(
        logcounts(sim_sub), gene_sets_sim, minSize = 2, kcdf = "Gaussian"
      ),
      BPPARAM = SerialParam()
    ),
    gsva(
      ssgseaParam(
        logcounts(sim_sub), gene_sets_sim, minSize = 2, normalize = FALSE
      ),
      BPPARAM = SerialParam()
    ),
    gsva(
      zscoreParam(logcounts(sim_sub), gene_sets_sim, minSize = 2),
      BPPARAM = SerialParam()
    ),
    gsva(
      plageParam(as.matrix(logcounts(sim_sub)), gene_sets_sim, minSize = 2),
      BPPARAM = SerialParam()
    ),
    genefunnel(
      counts(sim_sub), geneIds(gene_sets_sim),
      BPPARAM = SerialParam()
    ),
    iterations = 5,
    check = FALSE,
    filter_gc = FALSE
  )
  gc()

  saveRDS(bench, file = rds)
}

sim_big <- cbind(
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub,
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub,
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub,
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub,
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub,
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub,
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub,
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub,
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub,
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub,
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub,
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub,
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub,
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub,
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub,
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub,
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub,
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub,
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub,
  sim_sub, sim_sub, sim_sub, sim_sub, sim_sub
)

rds <- file.path(cache_dir, "bench_big.rds")
if (file.exists(rds)) {
  bench_big <- readRDS(rds)
} else {

  cores <- availableCores()
  gc()
  bench_big <- mark(
    gsva(
      gsvaParam(
        counts(sim_big), gene_sets_sim, minSize = 2, kcdf = "Poisson"
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      gsvaParam(
        logcounts(sim_big), gene_sets_sim, minSize = 2, kcdf = "Gaussian"
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      ssgseaParam(
        logcounts(sim_big), gene_sets_sim, minSize = 2, normalize = FALSE
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      zscoreParam(logcounts(sim_big), gene_sets_sim, minSize = 2),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      plageParam(as.matrix(logcounts(sim_big)), gene_sets_sim, minSize = 2),
      BPPARAM = MulticoreParam(cores)
    ),
    genefunnel(
      counts(sim_big), geneIds(gene_sets_sim),
      BPPARAM = MulticoreParam(cores)
    ),
    iterations = 5,
    check = FALSE,
    memory = FALSE,
    filter_gc = FALSE
  )
  gc()

  saveRDS(bench_big, file = rds)
}

sc <- counts(sc[ , which(sc$Group == "Group 1")])
set.seed(1)
sc <- sc[ , sample(ncol(sc), size = 20)]

tmp <- bplapply(
  seq_len(ncol(sc)),
  function(i) adjust_variability(sc[, i], high_var = FALSE),
  BPPARAM = MulticoreParam(availableCores())
)
low_var <- do.call(cbind, tmp)
tmp <- bplapply(
  seq_len(ncol(sc)),
  function(i) adjust_variability(sc[, i], high_var = TRUE),
  BPPARAM = MulticoreParam(availableCores())
)
high_var <- do.call(cbind, tmp)

sc <- cbind(cbind(sc, low_var), high_var)
sc <- SingleCellExperiment(list(counts = sc))
sc$Group <- factor(
  c(
    rep("Group1", times = ncol(low_var)),
    rep("Group2", times = ncol(low_var)),
    rep("Group3", times = ncol(low_var))
  )
)
logcounts(sc) <- Matrix(log2(counts(sc) + 1), sparse = TRUE)

rds <- file.path(cache_dir, "out_sc.rds")
if (file.exists(rds)) {
  out_sc <- readRDS(rds)
} else {

  out_sc <- vector("list", length = 6)
  names(out_sc) <- c(
    "GSVA Poisson", "GSVA Gaussian", "ssGSEA", "PLAGE", "Z-score", "GeneFunnel"
  )

  tmp <- gsvaParam(
    counts(sc), gene_sets_sim, minSize = 2, kcdf = "Poisson"
  )
  gc()
  out_sc[[1]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

  tmp <- gsvaParam(
    logcounts(sc), gene_sets_sim, minSize = 2, kcdf = "Gaussian"
  )
  gc()
  out_sc[[2]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

  tmp <- ssgseaParam(
    logcounts(sc), gene_sets_sim, minSize = 2, normalize = FALSE
  )
  gc()
  out_sc[[3]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

  tmp <- zscoreParam(
    logcounts(sc), gene_sets_sim, minSize = 2
  )
  gc()
  out_sc[[4]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

  tmp <- plageParam(
    as.matrix(logcounts(sc)), gene_sets_sim, minSize = 2,
  )
  gc()
  out_sc[[5]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

  gc()
  out_sc[[6]] <- genefunnel(
    counts(sc), geneIds(gene_sets_sim),
    BPPARAM = MulticoreParam(availableCores())
  )
  # out_sc[[6]] <- pmax(out_sc[[6]], 0)

  saveRDS(out_sc, file = rds)
}

rds <- file.path(cache_dir, "bench_sc.rds")
if (file.exists(rds)) {
  bench_sc <- readRDS(rds)
} else {

  cores <- availableCores()
  gc()
  bench_sc <- mark(
    gsva(
      gsvaParam(
        counts(sc), gene_sets_sim, minSize = 2, kcdf = "Poisson"
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      gsvaParam(
        logcounts(sc), gene_sets_sim, minSize = 2, kcdf = "Gaussian"
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      ssgseaParam(
        logcounts(sc), gene_sets_sim, minSize = 2, normalize = FALSE
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      zscoreParam(logcounts(sc), gene_sets_sim, minSize = 2),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      plageParam(as.matrix(logcounts(sc)), gene_sets_sim, minSize = 2),
      BPPARAM = MulticoreParam(cores)
    ),
    genefunnel(
      counts(sc), geneIds(gene_sets_sim),
      BPPARAM = MulticoreParam(cores)
    ),
    iterations = 5,
    check = FALSE,
    memory = FALSE,
    filter_gc = FALSE
  )
  gc()

  saveRDS(bench_sc, file = rds)
}

sc_big <- cbind(sc, sc)

rds <- file.path(cache_dir, "bench_sc2.rds")
if (file.exists(rds)) {
  bench_sc2 <- readRDS(rds)
} else {

  cores <- availableCores()
  gc()
  bench_sc2 <- mark(
    gsva(
      gsvaParam(
        counts(sc_big), gene_sets_sim, minSize = 2, kcdf = "Poisson"
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      gsvaParam(
        logcounts(sc_big), gene_sets_sim, minSize = 2, kcdf = "Gaussian"
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      ssgseaParam(
        logcounts(sc_big), gene_sets_sim, minSize = 2, normalize = FALSE
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      zscoreParam(logcounts(sc_big), gene_sets_sim, minSize = 2),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      plageParam(as.matrix(logcounts(sc_big)), gene_sets_sim, minSize = 2),
      BPPARAM = MulticoreParam(cores)
    ),
    genefunnel(
      counts(sc_big), geneIds(gene_sets_sim),
      BPPARAM = MulticoreParam(cores)
    ),
    iterations = 5,
    check = FALSE,
    memory = FALSE,
    filter_gc = FALSE
  )
  gc()

  saveRDS(bench_sc2, file = rds)
}

sc_big <- cbind(sc_big, sc_big)

rds <- file.path(cache_dir, "bench_sc3.rds")
if (file.exists(rds)) {
  bench_sc3 <- readRDS(rds)
} else {

  cores <- availableCores()
  gc()
  bench_sc3 <- mark(
    gsva(
      gsvaParam(
        counts(sc_big), gene_sets_sim, minSize = 2, kcdf = "Poisson"
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      gsvaParam(
        logcounts(sc_big), gene_sets_sim, minSize = 2, kcdf = "Gaussian"
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      ssgseaParam(
        logcounts(sc_big), gene_sets_sim, minSize = 2, normalize = FALSE
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      zscoreParam(logcounts(sc_big), gene_sets_sim, minSize = 2),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      plageParam(as.matrix(logcounts(sc_big)), gene_sets_sim, minSize = 2),
      BPPARAM = MulticoreParam(cores)
    ),
    genefunnel(
      counts(sc_big), geneIds(gene_sets_sim),
      BPPARAM = MulticoreParam(cores)
    ),
    iterations = 5,
    check = FALSE,
    memory = FALSE,
    filter_gc = FALSE
  )
  gc()

  saveRDS(bench_sc3, file = rds)
}

sc_big <- cbind(sc_big, sc_big)

rds <- file.path(cache_dir, "bench_sc4.rds")
if (file.exists(rds)) {
  bench_sc4 <- readRDS(rds)
} else {

  cores <- availableCores()
  gc()
  bench_sc4 <- mark(
    gsva(
      gsvaParam(
        counts(sc_big), gene_sets_sim, minSize = 2, kcdf = "Poisson"
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      gsvaParam(
        logcounts(sc_big), gene_sets_sim, minSize = 2, kcdf = "Gaussian"
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      ssgseaParam(
        logcounts(sc_big), gene_sets_sim, minSize = 2, normalize = FALSE
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      zscoreParam(logcounts(sc_big), gene_sets_sim, minSize = 2),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      plageParam(as.matrix(logcounts(sc_big)), gene_sets_sim, minSize = 2),
      BPPARAM = MulticoreParam(cores)
    ),
    genefunnel(
      counts(sc_big), geneIds(gene_sets_sim),
      BPPARAM = MulticoreParam(cores)
    ),
    iterations = 5,
    check = FALSE,
    memory = FALSE,
    filter_gc = FALSE
  )
  gc()

  saveRDS(bench_sc4, file = rds)
}

sc_big <- cbind(sc_big, sc_big)

rds <- file.path(cache_dir, "bench_sc5.rds")
if (file.exists(rds)) {
  bench_sc5 <- readRDS(rds)
} else {

  cores <- availableCores()
  gc()
  bench_sc5 <- mark(
    gsva(
      gsvaParam(
        counts(sc_big), gene_sets_sim, minSize = 2, kcdf = "Poisson"
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      gsvaParam(
        logcounts(sc_big), gene_sets_sim, minSize = 2, kcdf = "Gaussian"
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      ssgseaParam(
        logcounts(sc_big), gene_sets_sim, minSize = 2, normalize = FALSE
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      zscoreParam(logcounts(sc_big), gene_sets_sim, minSize = 2),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      plageParam(as.matrix(logcounts(sc_big)), gene_sets_sim, minSize = 2),
      BPPARAM = MulticoreParam(cores)
    ),
    genefunnel(
      counts(sc_big), geneIds(gene_sets_sim),
      BPPARAM = MulticoreParam(cores)
    ),
    iterations = 5,
    check = FALSE,
    memory = FALSE,
    filter_gc = FALSE
  )
  gc()

  saveRDS(bench_sc5, file = rds)
}

sc_big <- cbind(sc_big, sc_big)

rds <- file.path(cache_dir, "bench_sc6.rds")
if (file.exists(rds)) {
  bench_sc6 <- readRDS(rds)
} else {

  cores <- availableCores()
  gc()
  bench_sc6 <- mark(
    gsva(
      gsvaParam(
        counts(sc_big), gene_sets_sim, minSize = 2, kcdf = "Poisson"
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      gsvaParam(
        logcounts(sc_big), gene_sets_sim, minSize = 2, kcdf = "Gaussian"
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      ssgseaParam(
        logcounts(sc_big), gene_sets_sim, minSize = 2, normalize = FALSE
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      zscoreParam(logcounts(sc_big), gene_sets_sim, minSize = 2),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      plageParam(as.matrix(logcounts(sc_big)), gene_sets_sim, minSize = 2),
      BPPARAM = MulticoreParam(cores)
    ),
    genefunnel(
      counts(sc_big), geneIds(gene_sets_sim),
      BPPARAM = MulticoreParam(cores)
    ),
    iterations = 5,
    check = FALSE,
    memory = FALSE,
    filter_gc = FALSE
  )
  gc()

  saveRDS(bench_sc6, file = rds)
}

sc_big <- cbind(sc_big, sc_big)

rds <- file.path(cache_dir, "bench_sc7.rds")
if (file.exists(rds)) {
  bench_sc7 <- readRDS(rds)
} else {

  cores <- availableCores()
  gc()
  bench_sc7 <- mark(
    gsva(
      gsvaParam(
        counts(sc_big), gene_sets_sim, minSize = 2, kcdf = "Poisson"
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      gsvaParam(
        logcounts(sc_big), gene_sets_sim, minSize = 2, kcdf = "Gaussian"
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      ssgseaParam(
        logcounts(sc_big), gene_sets_sim, minSize = 2, normalize = FALSE
      ),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      zscoreParam(logcounts(sc_big), gene_sets_sim, minSize = 2),
      BPPARAM = MulticoreParam(cores)
    ),
    gsva(
      plageParam(as.matrix(logcounts(sc_big)), gene_sets_sim, minSize = 2),
      BPPARAM = MulticoreParam(cores)
    ),
    genefunnel(
      counts(sc_big), geneIds(gene_sets_sim),
      BPPARAM = MulticoreParam(cores)
    ),
    iterations = 5,
    check = FALSE,
    memory = FALSE,
    filter_gc = FALSE
  )
  gc()

  saveRDS(bench_sc7, file = rds)
}
```

Algorithm Overview
===

Row {data-height=355}
---

### About GeneFunnel 1 {.no-title}

GeneFunnel is a tool for gene set enrichment (or functional class scoring to be more precise).
It takes as input a matrix of genes/proteins (generally referred to as features) and samples, for example the output of an (sc)RNAseq or mass-spectrometry proteomics experiment.
Also required is a gene set list, such as from the Gene Ontology (GO), where each element is a set of genes with a name for each set.

$$
\text{score}_{k,j} = \sum_{i \in G_k} X_{i,j} - \left( \frac{|G_k|}{2(|G_k| - 1)} \sum_{i \in G_k} \left| X_{i,j} - \bar{X}_{G_k,j} \right| \right)
$$
Where:

- $\sum_{i \in G_k} X_{i,j}$ is the sum of the expression levels for the features in gene set $G_k$ for sample $j$.
- $\bar{X}_{G_k,j}$ is the mean expression of the features in gene set $G_k$ for sample $j$.
- $\sum_{i \in G_k} \left| X_{i,j} - \bar{X}_{G_k,j} \right|$ is the sum of the absolute deviations from the mean.
- $\frac{|G_k|}{2(|G_k| - 1)}$ is the scaling factor, which accounts for the number of features in the gene set and adjusts the influence of deviation.

### About GeneFunnel 2 {.no-title}

GeneFunnel iterates through samples in the matrix, where for each sample, the feature list is iterated through.
For each set in the list, the matrix is subset to the matching features and scored.
Conceptually, scoring takes the sum of the matching features, then adjusts the sum based on the deviance of each feature from the mean of the matching features.
Specifically, the measure of deviance is taken by subtracting the value of each feature from the mean of matched features, taking the absolute value, then summing all of the deviance values.
The absolute value is necessary as many features will have values lower than the mean, and we are interested in any deviance whether positive or negative.
  
A scaling factor is then derived, defined as the size of the gene set divided by twice the gene set size after an initial subtraction of 1.
Finally for each gene set, the summed deviance is multiplied by the scaling factor and then subtracted from the sum, yielding the score.
This provides the property that a minimally deviant set will be equal to the sum, while a maximally deviant set will be equal to zero.
Furthermore, a larger gene set with the same mean and deviance as a smaller one will have a greater score.

Row {data-height=650}
---

### Interactive Example {.no-title}

> Below is a contrived scenario of genes and samples with the resulting GeneFunnel scoring on the right and associated metrics.
> Try editing some values to see it update in real-time!

```{r}
shinyApp(
  ui = fluidPage(
    mainPanel(
      width = 4,
      tags$h3(
        HTML("тож Click Cells to Edit Matrix тоз"), style = "text-align:center"
      ),
      matrixInput(
        "mat",
        label = "Hypothetical Raw Gene Count by Sample Matrix",
        value = default_mat
      )
    ),
    mainPanel(
      width = 3,
      plotOutput("gene_heatmap", height = "400px")
    ),
    mainPanel(
      width = 5,
      plotOutput("gse_heatmap", height = "400px")
    )
  ),
  server = function(input, output, session) {

    output$gene_heatmap <- renderPlot({
      ht_opt$DIMNAME_PADDING <- unit(3, units = "mm")

      heatmap_mat <- apply(input$mat, MARGIN = 2, FUN = as.numeric)
      rownames(heatmap_mat) <- rownames(input$mat)

      col_range <- c(
        min(heatmap_mat, na.rm = TRUE),
        (min(heatmap_mat, na.rm = TRUE) + max(heatmap_mat, na.rm = TRUE)) / 2,
        max(heatmap_mat, na.rm = TRUE)
      )

      draw(
        Heatmap(
          heatmap_mat,
          name = "Raw Gene Counts",
          col = colorRamp2(col_range, colors = c("blue", "white", "red")),
          cluster_rows = FALSE,
          cluster_columns = FALSE,
          show_column_names = FALSE,
          top_annotation = top_anno7,
          column_split = seq(5),
          column_title = "Hypothetical Raw Gene Count by Sample Matrix",
          rect_gp = gpar(col = "black", lwd = 0.5),
          column_gap = unit(1, units = "mm"),
          row_names_gp = gpar(fontsize = 14),
          column_title_gp = gpar(fontsize = 13),
          row_names_side = "left",
          heatmap_legend_param = list(
            labels_gp = gpar(fontsize = 12), title_gp = gpar(fontsize = 14),
            border = "lightgrey", title_position = "topcenter",
            grid_height = unit(3, units = "mm"), direction = "horizontal"
          )
        ),
        heatmap_legend_side = "top"
      )
    })

    output$gse_heatmap <- renderPlot({
      ht_opt$DIMNAME_PADDING <- unit(3, units = "mm")

      numeric_mat <- apply(input$mat, MARGIN = 2, FUN = as.numeric)
      heatmap_mat <- matrix(
        c(
          sum(numeric_mat[ , 1], na.rm = TRUE) -
            (sum(
              abs(numeric_mat[ , 1] - mean(numeric_mat[ , 1], na.rm = TRUE)),
              na.rm = TRUE
            ) * (10 / 18)),
          sum(numeric_mat[ , 2], na.rm = TRUE) -
            (sum(
              abs(numeric_mat[ , 2] - mean(numeric_mat[ , 2], na.rm = TRUE)),
              na.rm = TRUE
            ) * (10 / 18)),
          sum(numeric_mat[ , 3], na.rm = TRUE) -
            (sum(
              abs(numeric_mat[ , 3] - mean(numeric_mat[ , 3], na.rm = TRUE)),
              na.rm = TRUE
            ) * (10 / 18)),
          sum(numeric_mat[ , 4], na.rm = TRUE) -
            (sum(
              abs(numeric_mat[ , 4] - mean(numeric_mat[ , 4], na.rm = TRUE)),
              na.rm = TRUE
            ) * (5 / 8)),
          sum(numeric_mat[ , 5], na.rm = TRUE) -
            (sum(
              abs(numeric_mat[ , 5] - mean(numeric_mat[ , 5], na.rm = TRUE)),
              na.rm = TRUE
            ) * (10 / 18))
        ),
        nrow =  1,
        ncol = 5,
        dimnames = list("Gene Set X  \n(Genes A to J)")
      )

      bottom_anno <- HeatmapAnnotation(
        text = anno_text(
          c(
            paste0(
              "\n\n\n\nSum = ", sum(numeric_mat[ , 1], na.rm = TRUE),
              "\nMean = ", mean(numeric_mat[ , 1], na.rm = TRUE),
              "\nDeviance = ", round(
                sum(
                  abs(
                    numeric_mat[ , 1] - mean(numeric_mat[ , 1], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                ) * (10 / 18),
                digits = 0
              ),
              "\nScore = ", round(sum(numeric_mat[ , 1], na.rm = TRUE) -
                (sum(
                  abs(
                    numeric_mat[ , 1] - mean(numeric_mat[ , 1], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                ) * (10 / 18)))
            ),
            paste0(
              "\n\n\n\nSum = ", sum(numeric_mat[ , 2], na.rm = TRUE),
              "\nMean = ", mean(numeric_mat[ , 2], na.rm = TRUE),
              "\nDeviance = ", round(
                sum(
                  abs(
                    numeric_mat[ , 2] - mean(numeric_mat[ , 2], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                ) * (10 / 18),
                digits = 0
              ),
              "\nScore = ", round(sum(numeric_mat[ , 2], na.rm = TRUE) -
                (sum(
                  abs(
                    numeric_mat[ , 2] - mean(numeric_mat[ , 2], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                ) * (10 / 18)))
            ),
            paste0(
              "\n\n\n\nSum = ", sum(numeric_mat[ , 3], na.rm = TRUE),
              "\nMean = ", mean(numeric_mat[ , 3], na.rm = TRUE),
               "\nDeviance = ", round(
                sum(
                  abs(
                    numeric_mat[ , 3] - mean(numeric_mat[ , 3], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                ) * (10 / 18),
                digits = 0
              ),
              "\nScore = ", round(sum(numeric_mat[ , 3], na.rm = TRUE) -
                (sum(
                  abs(
                    numeric_mat[ , 3] - mean(numeric_mat[ , 3], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                ) * (10 / 18)))
            ),
            paste0(
              "\n\n\n\nSum = ", sum(numeric_mat[ , 4], na.rm = TRUE),
              "\nMean = ", mean(numeric_mat[ , 4], na.rm = TRUE),
              "\nDeviance = ", round(
                sum(
                  abs(
                    numeric_mat[ , 4] - mean(numeric_mat[ , 4], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                ) * (5 / 8),
                digits = 0
              ),
              "\nScore = ", round(sum(numeric_mat[ , 4], na.rm = TRUE) -
                (sum(
                  abs(
                    numeric_mat[ , 4] - mean(numeric_mat[ , 4], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                ) * (5 / 8)))
            ),
            paste0(
              "\n\n\n\nSum = ", sum(numeric_mat[ , 5], na.rm = TRUE),
              "\nMean = ", mean(numeric_mat[ , 5], na.rm = TRUE),
              "\nDeviance = ", round(
                sum(
                  abs(
                    numeric_mat[ , 5] - mean(numeric_mat[ , 5], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                ) * (10 / 18),
                digits = 0
              ),
              "\nScore = ", round(sum(numeric_mat[ , 5], na.rm = TRUE) -
                (sum(
                  abs(
                    numeric_mat[ , 5] - mean(numeric_mat[ , 5], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                ) * (10 / 18)))
            )
          ),
          rot = 0,
          gp = gpar(fontsize = 16),
          just = "centre"
        )
      )

      col_range <- c(
        min(heatmap_mat, na.rm = TRUE),
        (min(heatmap_mat, na.rm = TRUE) + max(heatmap_mat, na.rm = TRUE)) / 2,
        max(heatmap_mat, na.rm = TRUE)
      )

      draw(
        Heatmap(
          heatmap_mat,
          name = "Gene Set Score",
          col = colorRamp2(col_range, colors = c("blue", "white", "red")),
          cluster_rows = FALSE,
          cluster_columns = FALSE,
          show_column_names = FALSE,
          top_annotation = top_anno,
          bottom_annotation = bottom_anno,
          column_split = seq(5),
          column_title = "Resulting Gene Set Enrichment by Sample Matrix",
          rect_gp = gpar(col = "black", lwd = 0.5),
          column_gap = unit(2.5, units = "mm"),
          row_names_gp = gpar(fontsize = 16),
          column_title_gp = gpar(fontsize = 18),
          row_names_side = "left",
          height = unit(0.5, units = "npc"),
          heatmap_legend_param = list(
            labels_gp = gpar(fontsize = 14), title_gp = gpar(fontsize = 15),
            border = "lightgrey", title_position = "topcenter",
            grid_height = unit(4, units = "mm"), direction = "horizontal"
          )
        ),
        heatmap_legend_side = "top",
        padding = unit(c(75, 2, 2, 2), units = "mm")
      )
    })
  }
)
```

Comparison with Other Methods
===

Row {data-height=500}
---

### Example Matrix {.no-title}

```{r}
mat <- default_mat[ , c(1:3, 5)]
mat[c(6, 7), 4] <- 100

col_range <- c(min(mat), (min(mat) + max(mat)) / 2, max(mat))
draw(
  Heatmap(
    mat,
    name = "Raw Gene Counts",
    col = colorRamp2(col_range, colors = c("blue", "white", "red")),
    cluster_rows = FALSE,
    cluster_columns = FALSE,
    show_column_names = FALSE,
    row_title = NULL,
    top_annotation = top_anno2,
    right_annotation = row_anno,
    column_split = seq(4),
    row_split = c(rep(1, times = 5), rep(2, times = 5)),
    column_title = "Hypothetical Raw Gene Count by Sample Matrix",
    rect_gp = gpar(col = "black", lwd = 0.5),
    column_gap = unit(1, units = "mm"),
    row_names_gp = gpar(fontsize = 15),
    column_title_gp = gpar(fontsize = 16),
    row_names_side = "left",
    heatmap_legend_param = list(
      labels_gp = gpar(fontsize = 14), title_gp = gpar(fontsize = 15),
      border = "lightgrey", title_position = "topcenter",
      grid_height = unit(4, units = "mm"), direction = "horizontal"
    )
  ),
  heatmap_legend_side = "top"
)

seekViewport("annotation_empty_1")
loc1 = deviceLoc(x = unit(0, "npc"), y = unit(0, "npc"))
seekViewport("annotation_empty_2")
loc2 = deviceLoc(x = unit(1, "npc"), y = unit(1, "npc"))

seekViewport("global")
grid.rect(
  loc2$x * 0.95, (loc1$y + loc2$y) / 2,
  width = loc2$x - loc1$x, height = loc2$y * 1.93,
  just = c("left", "centre")
)
grid.text(
  "Gene Set Z", rot = 90,
  x = (loc1$x + loc2$x) * 0.501, y = (loc1$y + loc2$y) * 0.5
)
```

### Background {.no-title}

> On this page, the hypothetical gene count matrix shown left is compared across 5 different single-sample GSEA methods and GeneFunnel.  
>  
> The gene count matrix is the same as the one in Algorithm Overview except the last sample is excluded because most of the methods cannot handle NaN values.  
>  
> Three types of gene sets are considered, as annotated on the right-hand side of the count matrix: the first 5 genes, the last 5 genes, and all 10 genes.  
>  
> Heatmaps for each method are shown in the bottom-left and to the right are the raw numerical output tables.

```{r}
```

Row {data-height=555}
---

### Gene Set Scoring {.no-title}

```{r, include = FALSE}
out <- vector("list", length = 6)
names(out) <- c(
  "GSVA Poisson", "GSVA Gaussian", "ssGSEA", "PLAGE", "Z-score", "GeneFunnel"
)

tmp <- gsvaParam(mat, gene_sets, minSize = 2, kcdf = "Poisson")
out[[1]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

tmp <- gsvaParam(log2(mat + 1), gene_sets, minSize = 2, kcdf = "Gaussian")
out[[2]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

tmp <- ssgseaParam(log2(mat + 1), gene_sets, minSize = 2, normalize = FALSE)
out[[3]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

tmp <- zscoreParam(log2(mat + 1), gene_sets, minSize = 2)
out[[4]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

tmp <- plageParam(log2(mat + 1), gene_sets, minSize = 2)
out[[5]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

out[[6]] <- genefunnel(
  mat, gene_sets, BPPARAM = MulticoreParam(availableCores())
)
```

```{r, fig.width = 7}
grid_list <- vector("list", length = 2)

ht_list = NULL
for (i in c(1:3)) {
  col_range <- c(
    min(out[[i]], na.rm = TRUE),
    (min(out[[i]], na.rm = TRUE) + max(out[[i]], na.rm = TRUE)) / 2,
    max(out[[i]], na.rm = TRUE)
  )
  ht_list <- ht_list +
    Heatmap(
      out[[i]],
      col = colorRamp2(col_range, colors = c("blue", "white", "red")),
      name = names(out)[[i]],
      cluster_rows = FALSE,
      cluster_columns = FALSE,
      show_column_names = FALSE,
      top_annotation = top_anno3,
      column_split = seq(4),
      column_title = NULL,
      rect_gp = gpar(col = "black", lwd = 0.5),
      column_gap = unit(1, units = "mm"),
      row_names_gp = gpar(fontsize = 16),
      column_title_gp = gpar(fontsize = 18),
      row_names_side = "left",
      width = unit(4, units = "cm"),
      height = unit(2, units = "cm"),
      heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 14), title_gp = gpar(fontsize = 15),
        border = "lightgrey", title_position = "topcenter",
        grid_height = unit(4, units = "mm"), direction = "horizontal",
        legend_width = unit(3.75, units = "cm")
      )
    )
}
grid_list[[1]] <- grid.grabExpr(
  draw(
    ht_list, heatmap_legend_side = "top", legend_gap = unit(0.75, units = "cm")
  )
)

ht_list = NULL
for (i in c(4:6)) {
  col_range <- c(
    min(out[[i]], na.rm = TRUE),
    (min(out[[i]], na.rm = TRUE) + max(out[[i]], na.rm = TRUE)) / 2,
    max(out[[i]], na.rm = TRUE)
  )
  ht_list <- ht_list +
    Heatmap(
      out[[i]],
      col = colorRamp2(col_range, colors = c("blue", "white", "red")),
      name = names(out)[[i]],
      cluster_rows = FALSE,
      cluster_columns = FALSE,
      show_column_names = FALSE,
      top_annotation = top_anno3,
      column_split = seq(4),
      column_title = NULL,
      rect_gp = gpar(col = "black", lwd = 0.5),
      column_gap = unit(1, units = "mm"),
      row_names_gp = gpar(fontsize = 16),
      column_title_gp = gpar(fontsize = 18),
      row_names_side = "left",
      width = unit(4, units = "cm"),
      height = unit(2, units = "cm"),
      heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 14), title_gp = gpar(fontsize = 15),
        border = "lightgrey", title_position = "topcenter",
        grid_height = unit(4, units = "mm"), direction = "horizontal",
        legend_width = unit(3.75, units = "cm")
      )
    )
}
grid_list[[2]] <- grid.grabExpr(
  draw(
    ht_list, heatmap_legend_side = "top", legend_gap = unit(0.75, units = "cm")
  )
)

plot_grid(grid_list[[1]], grid_list[[2]], nrow = 2)
```

### Scoring Output 1 {.no-title}

```{r, results = "asis"}
for (i in c(1:3)) {
  print(
    kable_styling(
      kable(out[[i]], col.names = colnames, caption = names(out)[i]),
      bootstrap_options = "none"
    )
  )
}
```

### Scoring Output 2 {.no-title}

```{r, results = "asis"}
for (i in c(4:6)) {
  print(
    kable_styling(
      kable(out[[i]], col.names = colnames, caption = names(out)[i]),
      bootstrap_options = "none"
    )
  )
}
```

Real Dataset Gene Set Enrichment
===

```{r}
ht_opt$DIMNAME_PADDING <- unit(2, units = "mm")
```

Row {.tabset}
---

### Background

> In this next section we work with real data, in this case single-cell RNAseq data from Alzheimer's Disease post-mortem brain tissue [Otero-Garcia, 2022, Neuron](https://doi.org/10.1016/j.neuron.2022.06.021).
> In brief, the authors used FACS sorting to separate neurons harboring neurofibrillary tangles (a major pathological hallmark of Alzheimer's Disease) from nearby neurons without tangles, within the same donors.  
>  
> We used a standard processing pipeline (https://github.com/eturkes/otero-garcia-2022-ssRNAseq) and merged cell-type populations described by the authors to have a large proportion of tangle-bearing neurons.
> In the UMAP below, this corresponds to the populations EX L2-4 CUX2, EX L4-5 RORB, EX L5 RORB, and EX L5-6 THEMIS.

<div>

```{r, out.height = "100%", out.widgth = "100%", dpi = 150}
include_graphics("otero_garcia_umap")
```

</div>

### Non-overlapping Gene Sets

For this example, two gene sets are altered.
**NELF Complex** is modified to reduce variability, that is, all genes are set to the sum of the gene counts divided by the total number of genes.
**Trace-amine Receptor Activity** is simply modified to have increased counts - all genes have 100 counts added to them.
These gene sets were specifically selected because they have no gene overlap with other gene sets in the testing set, therefore, any changes detected should only be in these two sets.
Below are the top two gene sets sorted by greatest absolute difference between the modified and the original column.

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r, fig.width = 8, fig.height = 7}
grid_list <- vector("list", length = 3)

for (i in c(1:3)) {
  mat <- SingleCellExperiment(list(counts = out_sim[[i]]))
  mat$Group <- sim$Group
  mat <- mat[ , c(1, 4)]
  col_range <- c(
    min(counts(mat)),
    (min(counts(mat)) + max(counts(mat))) / 2,
    max(counts(mat))
  )

  tmp <- mat[order(counts(mat[ , 2]) - counts(mat[ , 1]), decreasing = TRUE), ]
  rownames <- rownames(tmp)[1:2]
  mat <- mat[rownames(mat) %in% rownames, ]
  mat <- mat[match(rownames, rownames(mat)), ]
  grid_list[[i]] <- Heatmap(
    counts(mat),
    name = names(out_sim)[[i]],
    col = colorRamp2(col_range, colors = c("blue", "white", "red")),
    cluster_rows = FALSE,
    cluster_columns = FALSE,
    show_column_names = FALSE,
    top_annotation = top_anno4,
    column_split = mat$Group,
    column_title = NULL,
    rect_gp = gpar(col = "black", lwd = 0.5),
    column_gap = unit(1, units = "mm"),
    row_names_gp = gpar(fontsize = 14),
    column_title_gp = gpar(fontsize = 16),
    row_names_side = "left",
    row_names_max_width = max_text_width(
      rownames(mat), gp = gpar(fontsize = 14)
    ),
    heatmap_legend_param = list(
      labels_gp = gpar(fontsize = 12), title_gp = gpar(fontsize = 14),
      border = "lightgrey", title_position = "topcenter",
      grid_height = unit(4, units = "mm"), direction = "horizontal"
    )
  )

  grid_list[[i]] <- grid.grabExpr(
    draw(grid_list[[i]], heatmap_legend_side = "top")
  )
}

plot_grid(grid_list[[1]], grid_list[[2]], grid_list[[3]], nrow = 3)
```

</div>
<div>

```{r, fig.width = 8, fig.height = 7}
grid_list <- vector("list", length = 3)

for (i in c(4:6)) {
  mat <- SingleCellExperiment(list(counts = out_sim[[i]]))
  mat$Group <- sim$Group
  mat <- mat[ , c(1, 4)]
  col_range <- c(
    min(counts(mat)),
    (min(counts(mat)) + max(counts(mat))) / 2,
    max(counts(mat))
  )

  tmp <- mat[order(counts(mat[ , 2]) - counts(mat[ , 1]), decreasing = TRUE), ]
  rownames <- rownames(tmp)[1:2]
  mat <- mat[rownames(mat) %in% rownames, ]
  mat <- mat[match(rownames, rownames(mat)), ]
  col_range <- c(
    min(counts(mat)),
    (min(counts(mat)) + max(counts(mat))) / 2,
    max(counts(mat))
  )
  grid_list[[i - 3]] <- Heatmap(
    counts(mat),
    name = names(out_sim)[[i]],
    col = colorRamp2(col_range, colors = c("blue", "white", "red")),
    cluster_rows = FALSE,
    cluster_columns = FALSE,
    show_column_names = FALSE,
    top_annotation = top_anno4,
    column_split = mat$Group,
    column_title = NULL,
    rect_gp = gpar(col = "black", lwd = 0.5),
    column_gap = unit(1, units = "mm"),
    row_names_gp = gpar(fontsize = 14),
    column_title_gp = gpar(fontsize = 16),
    row_names_side = "left",
    row_names_max_width = max_text_width(
      rownames(mat), gp = gpar(fontsize = 14)
    ),
    heatmap_legend_param = list(
      labels_gp = gpar(fontsize = 12), title_gp = gpar(fontsize = 14),
      border = "lightgrey", title_position = "topcenter",
      grid_height = unit(4, units = "mm"), direction = "horizontal"
    )
  )

  grid_list[[i - 3]] <- grid.grabExpr(
    draw(grid_list[[i - 3]], heatmap_legend_side = "top")
  )
}

plot_grid(grid_list[[1]], grid_list[[2]], grid_list[[3]], nrow = 3)
```

</div>
</div>

### Tau Protein Binding

**Tau Protein Binding** is modified to reduce variability, that is, all genes are set to the sum of the gene counts divided by the total number of genes.
Note that this gene set has overlap with other gene sets, so it may not necessarily be the highest hit, but should be ranked near the top.
Below are the top 5 gene sets sorted by greatest absolute difference between the modified and original column.

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r, fig.width = 8, fig.height = 7}
grid_list <- vector("list", length = 3)

for (i in c(1:3)) {
  mat <- SingleCellExperiment(list(counts = out_sim[[i]]))
  mat$Group <- sim$Group
  mat <- mat[ , c(1, 5)]
  col_range <- c(
    min(counts(mat)),
    (min(counts(mat)) + max(counts(mat))) / 2,
    max(counts(mat))
  )

  tmp <- mat[order(counts(mat[ , 2]) - counts(mat[ , 1]), decreasing = TRUE), ]
  rownames <- rownames(tmp)[1:5]
  mat <- mat[rownames(mat) %in% rownames, ]
  mat <- mat[match(rownames, rownames(mat)), ]
  grid_list[[i]] <- Heatmap(
    counts(mat),
    name = names(out_sim)[[i]],
    col = colorRamp2(col_range, colors = c("blue", "white", "red")),
    cluster_rows = FALSE,
    cluster_columns = FALSE,
    show_column_names = FALSE,
    top_annotation = top_anno4,
    column_split = mat$Group,
    column_title = NULL,
    rect_gp = gpar(col = "black", lwd = 0.5),
    column_gap = unit(1, units = "mm"),
    row_names_gp = gpar(fontsize = 14),
    column_title_gp = gpar(fontsize = 16),
    row_names_side = "left",
    row_names_max_width = max_text_width(
      rownames(mat), gp = gpar(fontsize = 14)
    ),
    heatmap_legend_param = list(
      labels_gp = gpar(fontsize = 12), title_gp = gpar(fontsize = 14),
      border = "lightgrey", title_position = "topcenter",
      grid_height = unit(4, units = "mm"), direction = "horizontal"
    )
  )

  grid_list[[i]] <- grid.grabExpr(
    draw(grid_list[[i]], heatmap_legend_side = "top")
  )
}

plot_grid(grid_list[[1]], grid_list[[2]], grid_list[[3]], nrow = 3)
```

</div>
<div>

```{r, fig.width = 8, fig.height = 7}
grid_list <- vector("list", length = 3)

for (i in c(4:6)) {
  mat <- SingleCellExperiment(list(counts = out_sim[[i]]))
  mat$Group <- sim$Group
  mat <- mat[ , c(1, 5)]
  col_range <- c(
    min(counts(mat)),
    (min(counts(mat)) + max(counts(mat))) / 2,
    max(counts(mat))
  )

  tmp <- mat[order(counts(mat[ , 2]) - counts(mat[ , 1]), decreasing = TRUE), ]
  rownames <- rownames(tmp)[1:5]
  mat <- mat[rownames(mat) %in% rownames, ]
  mat <- mat[match(rownames, rownames(mat)), ]
  grid_list[[i - 3]] <- Heatmap(
    counts(mat),
    name = names(out_sim)[[i]],
    col = colorRamp2(col_range, colors = c("blue", "white", "red")),
    cluster_rows = FALSE,
    cluster_columns = FALSE,
    show_column_names = FALSE,
    top_annotation = top_anno4,
    column_split = mat$Group,
    column_title = NULL,
    rect_gp = gpar(col = "black", lwd = 0.5),
    column_gap = unit(1, units = "mm"),
    row_names_gp = gpar(fontsize = 14),
    column_title_gp = gpar(fontsize = 16),
    row_names_side = "left",
    row_names_max_width = max_text_width(
      rownames(mat), gp = gpar(fontsize = 14)
    ),
    heatmap_legend_param = list(
      labels_gp = gpar(fontsize = 12), title_gp = gpar(fontsize = 14),
      border = "lightgrey", title_position = "topcenter",
      grid_height = unit(4, units = "mm"), direction = "horizontal"
    )
  )

  grid_list[[i - 3]] <- grid.grabExpr(
    draw(grid_list[[i - 3]], heatmap_legend_side = "top")
  )
}

plot_grid(grid_list[[1]], grid_list[[2]], grid_list[[3]], nrow = 3)
```

</div>
</div>

### Neurofibrillary Tangle

**Neurofibrillary Tangle** is modified to have increased counts - all genes have 100 counts added to them.
Note that this gene set has overlap with other gene sets, so it may not necessarily be the highest hit, but should be ranked near the top.
Below are the top 5 gene sets sorted by greatest absolute difference between the modified and original column.

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r, fig.width = 8, fig.height = 7}
grid_list <- vector("list", length = 3)

for (i in c(1:3)) {
  mat <- SingleCellExperiment(list(counts = out_sim[[i]]))
  mat$Group <- sim$Group
  mat <- mat[ , c(1, 6)]
  col_range <- c(
    min(counts(mat)),
    (min(counts(mat)) + max(counts(mat))) / 2,
    max(counts(mat))
  )

  tmp <- mat[order(counts(mat[ , 2]) - counts(mat[ , 1]), decreasing = TRUE), ]
  rownames <- rownames(tmp)[1:5]
  mat <- mat[rownames(mat) %in% rownames, ]
  mat <- mat[match(rownames, rownames(mat)), ]
  grid_list[[i]] <- Heatmap(
    counts(mat),
    name = names(out_sim)[[i]],
    col = colorRamp2(col_range, colors = c("blue", "white", "red")),
    cluster_rows = FALSE,
    cluster_columns = FALSE,
    show_column_names = FALSE,
    top_annotation = top_anno4,
    column_split = mat$Group,
    column_title = NULL,
    rect_gp = gpar(col = "black", lwd = 0.5),
    column_gap = unit(1, units = "mm"),
    row_names_gp = gpar(fontsize = 14),
    column_title_gp = gpar(fontsize = 16),
    row_names_side = "left",
    row_names_max_width = max_text_width(
      rownames(mat), gp = gpar(fontsize = 14)
    ),
    heatmap_legend_param = list(
      labels_gp = gpar(fontsize = 12), title_gp = gpar(fontsize = 14),
      border = "lightgrey", title_position = "topcenter",
      grid_height = unit(4, units = "mm"), direction = "horizontal"
    )
  )

  grid_list[[i]] <- grid.grabExpr(
    draw(grid_list[[i]], heatmap_legend_side = "top")
  )
}

plot_grid(grid_list[[1]], grid_list[[2]], grid_list[[3]], nrow = 3)
```

</div>
<div>

```{r, fig.width = 8, fig.height = 7}
grid_list <- vector("list", length = 3)

for (i in c(4:6)) {
  if (i == 5) {
    mat <- SingleCellExperiment(list(counts = out_sim[[i]]))
    mat$Group <- sim$Group
    mat <- mat[ , c(1, 6)]
    col_range <- c(
      min(counts(mat)),
      (min(counts(mat)) + max(counts(mat))) / 2,
      max(counts(mat))
    )

    tmp <- mat[
      order(counts(mat[ , 2]) - counts(mat[ , 1]), decreasing = TRUE),
    ]
    rownames <- rownames(tmp)[1:5]
    mat <- mat[rownames(mat) %in% rownames, ]
    mat <- mat[match(rownames, rownames(mat)), ]
    grid_list[[i - 3]] <- Heatmap(
      counts(mat),
      name = names(out_sim)[[i]],
      col = colorRamp2(col_range, colors = c("blue", "white", "red")),
      cluster_rows = FALSE,
      cluster_columns = FALSE,
      show_column_names = FALSE,
      top_annotation = top_anno8,
      column_split = mat$Group,
      column_title = NULL,
      rect_gp = gpar(col = "black", lwd = 0.5),
      column_gap = unit(1, units = "mm"),
      row_names_gp = gpar(fontsize = 10),
      column_title_gp = gpar(fontsize = 16),
      row_names_side = "left",
      row_names_max_width = max_text_width(
        rownames(mat), gp = gpar(fontsize = 10)
      ),
      heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 12), title_gp = gpar(fontsize = 14),
        border = "lightgrey", title_position = "topcenter",
        grid_height = unit(4, units = "mm"), direction = "horizontal"
      )
    )

    grid_list[[i - 3]] <- grid.grabExpr(
      draw(grid_list[[i - 3]], heatmap_legend_side = "top")
    )
  } else {
    mat <- SingleCellExperiment(list(counts = out_sim[[i]]))
    mat$Group <- sim$Group
    mat <- mat[ , c(1, 6)]
    col_range <- c(
      min(counts(mat)),
      (min(counts(mat)) + max(counts(mat))) / 2,
      max(counts(mat))
    )

    tmp <- mat[
      order(counts(mat[ , 2]) - counts(mat[ , 1]), decreasing = TRUE),
    ]
    rownames <- rownames(tmp)[1:5]
    mat <- mat[rownames(mat) %in% rownames, ]
    mat <- mat[match(rownames, rownames(mat)), ]
    grid_list[[i - 3]] <- Heatmap(
      counts(mat),
      name = names(out_sim)[[i]],
      col = colorRamp2(col_range, colors = c("blue", "white", "red")),
      cluster_rows = FALSE,
      cluster_columns = FALSE,
      show_column_names = FALSE,
      top_annotation = top_anno4,
      column_split = mat$Group,
      column_title = NULL,
      rect_gp = gpar(col = "black", lwd = 0.5),
      column_gap = unit(1, units = "mm"),
      row_names_gp = gpar(fontsize = 14),
      column_title_gp = gpar(fontsize = 16),
      row_names_side = "left",
      row_names_max_width = max_text_width(
        rownames(mat), gp = gpar(fontsize = 14)
      ),
      heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 12), title_gp = gpar(fontsize = 14),
        border = "lightgrey", title_position = "topcenter",
        grid_height = unit(4, units = "mm"), direction = "horizontal"
      )
    )

    grid_list[[i - 3]] <- grid.grabExpr(
      draw(grid_list[[i - 3]], heatmap_legend_side = "top")
    )
  }
}

plot_grid(grid_list[[1]], grid_list[[2]], grid_list[[3]], nrow = 3)
```

</div>
</div>

### Low vs. High Variability

In this section, the original column is modified producing columns with either lower or higher variability.
More specifically, the lower variability column will have values more closely centered around the mean, while higher will have a greater spread of values.
In both cases, the modification is done in a way that does not change the rank between genes as well as retains their relative distances from each other in terms of expression.
This helps keep enriched gene sets comparable between the columns.

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r, fig.width = 8, fig.height = 7}
mat <- SingleCellExperiment(list(counts = out_sim[[6]] + 1))
mat$Group <- sim$Group
mat <- mat[ , c(1, 2, 3)]
col_range <- c(
  min(counts(mat)),
  (min(counts(mat)) + max(counts(mat))) / 2,
  max(counts(mat))
)

# tmp <- mat[order(counts(mat[ , 3]), decreasing = TRUE), ]
tmp <- mat[order(counts(mat[ , 3]) - counts(mat[ , 2]), decreasing = TRUE), ]
rownames <- rownames(tmp)[c(1:10)]
tmp <- mat[order(counts(mat[ , 2]) - counts(mat[ , 3]), decreasing = TRUE), ]
rownames <- c(rownames, rownames(tmp)[c(1:10)])
# rownames <- rownames(tmp)[c(1:5, 4867:4872, 9733:9738)]
# rownames <- rownames(tmp)[c(1:10, 9728:9738)]
mat <- mat[rownames(mat) %in% rownames, ]
mat <- mat[match(rownames, rownames(mat)), ]
draw(
  Heatmap(
    counts(mat),
    name = names(out_sim)[[6]],
    col = colorRamp2(col_range, colors = c("blue", "white", "red")),
    cluster_rows = FALSE,
    cluster_columns = FALSE,
    show_column_names = FALSE,
    top_annotation = top_anno5,
    column_split = mat$Group,
    column_title = NULL,
    rect_gp = gpar(col = "black", lwd = 0.5),
    column_gap = unit(1, units = "mm"),
    row_names_gp = gpar(fontsize = 12),
    column_title_gp = gpar(fontsize = 16),
    row_names_side = "left",
    row_names_max_width = max_text_width(
      rownames(mat), gp = gpar(fontsize = 12)
    ),
    heatmap_legend_param = list(
      labels_gp = gpar(fontsize = 12), title_gp = gpar(fontsize = 14),
      border = "lightgrey", title_position = "topcenter",
      grid_height = unit(4, units = "mm"), direction = "horizontal"
    )
  ),
  heatmap_legend_side = "top"
)
```

</div>
<div>

```{r, fig.width = 8, fig.height = 7}
counts(mat) <- t(
  apply(
    counts(mat), MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
draw(
  Heatmap(
    counts(mat),
    name = paste(names(out_sim)[[6]], "Scaled", sep = " "),
    col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
    cluster_rows = FALSE,
    cluster_columns = FALSE,
    show_column_names = FALSE,
    top_annotation = top_anno5,
    column_split = mat$Group,
    column_title = NULL,
    rect_gp = gpar(col = "black", lwd = 0.5),
    column_gap = unit(1, units = "mm"),
    row_names_gp = gpar(fontsize = 12),
    column_title_gp = gpar(fontsize = 16),
    row_names_side = "left",
    row_names_max_width = max_text_width(
      rownames(mat), gp = gpar(fontsize = 12)
    ),
    heatmap_legend_param = list(
      labels_gp = gpar(fontsize = 12), title_gp = gpar(fontsize = 14),
      border = "lightgrey", title_position = "topcenter",
      grid_height = unit(4, units = "mm"), direction = "horizontal"
    )
  ),
  heatmap_legend_side = "top"
)
```

</div>
</div>

### Differential Expression (NFT vs. CTRL)

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r, fig.width = 8, fig.height = 7}
grid_list <- vector("list", length = 3)

for (i in c(1:3)) {
  mat <- SingleCellExperiment(list(counts = out_sim[[i]]))
  mat$Group <- sim$Group
  mat <- mat[ , c(7, 8)]
  col_range <- c(
    min(counts(mat)),
    (min(counts(mat)) + max(counts(mat))) / 2,
    max(counts(mat))
  )

  tmp <- mat[order(counts(mat[ , 1]) - counts(mat[ , 2]), decreasing = TRUE), ]
  rownames <- rownames(tmp)[1:5]
  mat <- mat[rownames(mat) %in% rownames, ]
  mat <- mat[match(rownames, rownames(mat)), ]
  grid_list[[i]] <- Heatmap(
    counts(mat),
    name = names(out_sim)[[i]],
    col = colorRamp2(col_range, colors = c("blue", "white", "red")),
    cluster_rows = FALSE,
    cluster_columns = FALSE,
    show_column_names = FALSE,
    top_annotation = top_anno6,
    column_split = mat$Group,
    column_title = NULL,
    rect_gp = gpar(col = "black", lwd = 0.5),
    column_gap = unit(1, units = "mm"),
    row_names_gp = gpar(fontsize = 14),
    column_title_gp = gpar(fontsize = 16),
    row_names_side = "left",
    row_names_max_width = max_text_width(
      rownames(mat), gp = gpar(fontsize = 14)
    ),
    heatmap_legend_param = list(
      labels_gp = gpar(fontsize = 12), title_gp = gpar(fontsize = 14),
      border = "lightgrey", title_position = "topcenter",
      grid_height = unit(4, units = "mm"), direction = "horizontal"
    )
  )

  grid_list[[i]] <- grid.grabExpr(
    draw(grid_list[[i]], heatmap_legend_side = "top")
  )
}

plot_grid(grid_list[[1]], grid_list[[2]], grid_list[[3]], nrow = 3)
```

</div>
<div>

```{r, fig.width = 8, fig.height = 7}
grid_list <- vector("list", length = 3)

for (i in c(4:6)) {
  mat <- SingleCellExperiment(list(counts = out_sim[[i]]))
  mat$Group <- sim$Group
  mat <- mat[ , c(7, 8)]
  col_range <- c(
    min(counts(mat)),
    (min(counts(mat)) + max(counts(mat))) / 2,
    max(counts(mat))
  )

  tmp <- mat[order(counts(mat[ , 1]) - counts(mat[ , 2]), decreasing = TRUE), ]
  rownames <- rownames(tmp)[1:5]
  mat <- mat[rownames(mat) %in% rownames, ]
  mat <- mat[match(rownames, rownames(mat)), ]
  grid_list[[i - 3]] <- Heatmap(
    counts(mat),
    name = names(out_sim)[[i]],
    col = colorRamp2(col_range, colors = c("blue", "white", "red")),
    cluster_rows = FALSE,
    cluster_columns = FALSE,
    show_column_names = FALSE,
    top_annotation = top_anno6,
    column_split = mat$Group,
    column_title = NULL,
    rect_gp = gpar(col = "black", lwd = 0.5),
    column_gap = unit(1, units = "mm"),
    row_names_gp = gpar(fontsize = 14),
    column_title_gp = gpar(fontsize = 16),
    row_names_side = "left",
    row_names_max_width = max_text_width(
      rownames(mat), gp = gpar(fontsize = 14)
    ),
    heatmap_legend_param = list(
      labels_gp = gpar(fontsize = 12), title_gp = gpar(fontsize = 14),
      border = "lightgrey", title_position = "topcenter",
      grid_height = unit(4, units = "mm"), direction = "horizontal"
    )
  )

  grid_list[[i - 3]] <- grid.grabExpr(
    draw(grid_list[[i - 3]], heatmap_legend_side = "top")
  )
}

plot_grid(grid_list[[1]], grid_list[[2]], grid_list[[3]], nrow = 3)
```

</div>
</div>

Real Dataset Gene Expression
===

Row {.tabset}
---

### Background

> This section compliments the previous section and shows the original gene expression values for the gene sets targeted.

```{r}
```

### Non-overlapping Gene Sets

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r, fig.width = 7, fig.height = 8}
gse_plot <- "NELF complex"
for (gse in gse_plot) {
  genes <- unlist(geneIds(gene_sets_sim[names(gene_sets_sim) %in% gse]))
  mat <- sim[rownames(sim) %in% genes, c(1, 4)]
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(mat), ]
  gene_anno_sub <- gene_anno_sub[match(rownames(mat), gene_anno_sub$ensembl), ]
  rownames(mat) <- gene_anno_sub$symbol
  col_range <- c(
    min(logcounts(mat)),
    (min(logcounts(mat)) + max(logcounts(mat))) / 2,
    max(logcounts(mat))
  )
  ht <- draw(
    Heatmap(
      logcounts(mat),
      name = "log2 + 1 Gene Expression",
      col = colorRamp2(col_range, colors = c("blue", "white", "red")),
      cluster_rows = FALSE,
      cluster_columns = FALSE,
      show_column_names = FALSE,
      top_annotation = top_anno4,
      column_split = mat$Group,
      column_title = gse,
      rect_gp = gpar(col = "black", lwd = 0.5),
      column_gap = unit(1, units = "mm"),
      row_names_gp = gpar(fontsize = 9),
      column_title_gp = gpar(fontsize = 12),
      row_names_side = "left",
      heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 10), title_gp = gpar(fontsize = 12),
        border = "lightgrey", title_position = "topcenter",
        grid_height = unit(4, "mm"), direction = "horizontal"
      )
    ),
    heatmap_legend_side = "top"
  )
  ht
}
```

</div>
<div>

```{r, fig.width = 7, fig.height = 8}
gse_plot <- "trace-amine receptor activity"
for (gse in gse_plot) {
  genes <- unlist(geneIds(gene_sets_sim[names(gene_sets_sim) %in% gse]))
  mat <- sim[rownames(sim) %in% genes, c(1, 4)]
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(mat), ]
  gene_anno_sub <- gene_anno_sub[match(rownames(mat), gene_anno_sub$ensembl), ]
  rownames(mat) <- gene_anno_sub$symbol
  col_range <- c(
    min(logcounts(mat)),
    (min(logcounts(mat)) + max(logcounts(mat))) / 2,
    max(logcounts(mat))
  )
  ht <- draw(
    Heatmap(
      logcounts(mat),
      name = "log2 + 1 Gene Expression",
      col = colorRamp2(col_range, colors = c("blue", "white", "red")),
      cluster_rows = FALSE,
      cluster_columns = FALSE,
      show_column_names = FALSE,
      top_annotation = top_anno4,
      column_split = mat$Group,
      column_title = gse,
      rect_gp = gpar(col = "black", lwd = 0.5),
      column_gap = unit(1, units = "mm"),
      row_names_gp = gpar(fontsize = 9),
      column_title_gp = gpar(fontsize = 12),
      row_names_side = "left",
      heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 10), title_gp = gpar(fontsize = 12),
        border = "lightgrey", title_position = "topcenter",
        grid_height = unit(4, "mm"), direction = "horizontal"
      )
    ),
    heatmap_legend_side = "top"
  )
  ht
}
```

</div>
</div>

### Tau Protein Binding

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r, fig.width = 7, fig.height = 8}
gse_plot <- "tau protein binding"
for (gse in gse_plot) {
  genes <- unlist(geneIds(gene_sets_sim[names(gene_sets_sim) %in% gse]))
  mat <- sim[rownames(sim) %in% genes, c(1, 5)]
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(mat), ]
  gene_anno_sub <- gene_anno_sub[match(rownames(mat), gene_anno_sub$ensembl), ]
  rownames(mat) <- gene_anno_sub$symbol
  col_range <- c(
    min(logcounts(mat)),
    (min(logcounts(mat)) + max(logcounts(mat))) / 2,
    max(logcounts(mat))
  )
  ht <- draw(
    Heatmap(
      logcounts(mat),
      name = "log2 + 1 Gene Expression",
      col = colorRamp2(col_range, colors = c("blue", "white", "red")),
      cluster_rows = FALSE,
      cluster_columns = FALSE,
      show_column_names = FALSE,
      top_annotation = top_anno4,
      column_split = mat$Group,
      column_title = gse,
      rect_gp = gpar(col = "black", lwd = 0.5),
      column_gap = unit(1, units = "mm"),
      row_names_gp = gpar(fontsize = 9),
      column_title_gp = gpar(fontsize = 12),
      row_names_side = "left",
      heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 10), title_gp = gpar(fontsize = 12),
        border = "lightgrey", title_position = "topcenter",
        grid_height = unit(4, "mm"), direction = "horizontal"
      )
    ),
    heatmap_legend_side = "top"
  )
  ht
}
```

</div>
<div>

```{r, fig.width = 7, fig.height = 8}
gse_plot <- "tau protein binding"
for (gse in gse_plot) {
  genes <- unlist(geneIds(gene_sets_sim[names(gene_sets_sim) %in% gse]))
  mat <- sim[rownames(sim) %in% genes, c(1, 5)]
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(mat), ]
  gene_anno_sub <- gene_anno_sub[match(rownames(mat), gene_anno_sub$ensembl), ]
  rownames(mat) <- gene_anno_sub$symbol
  counts(mat) <- t(
    apply(
      counts(mat), MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  ht <- draw(
    Heatmap(
      counts(mat),
      name = "Scaled Gene Expression",
      col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
      cluster_rows = FALSE,
      cluster_columns = FALSE,
      show_column_names = FALSE,
      top_annotation = top_anno4,
      column_split = mat$Group,
      column_title = gse,
      rect_gp = gpar(col = "black", lwd = 0.5),
      column_gap = unit(1, units = "mm"),
      row_names_gp = gpar(fontsize = 9),
      column_title_gp = gpar(fontsize = 12),
      row_names_side = "left",
      heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 10), title_gp = gpar(fontsize = 12),
        border = "lightgrey", title_position = "topcenter",
        grid_height = unit(4, "mm"), direction = "horizontal"
      )
    ),
    heatmap_legend_side = "top"
  )
  ht
}
```

</div>
</div>

### Neurofibrillary Tangle

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r, fig.width = 7, fig.height = 8}
gse_plot <- "neurofibrillary tangle"
for (gse in gse_plot) {
  genes <- unlist(geneIds(gene_sets_sim[names(gene_sets_sim) %in% gse]))
  mat <- sim[rownames(sim) %in% genes, c(1, 6)]
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(mat), ]
  gene_anno_sub <- gene_anno_sub[match(rownames(mat), gene_anno_sub$ensembl), ]
  rownames(mat) <- gene_anno_sub$symbol
  col_range <- c(
    min(logcounts(mat)),
    (min(logcounts(mat)) + max(logcounts(mat))) / 2,
    max(logcounts(mat))
  )
  ht <- draw(
    Heatmap(
      logcounts(mat),
      name = "log2 + 1 Gene Expression",
      col = colorRamp2(col_range, colors = c("blue", "white", "red")),
      cluster_rows = FALSE,
      cluster_columns = FALSE,
      show_column_names = FALSE,
      top_annotation = top_anno4,
      column_split = mat$Group,
      column_title = gse,
      rect_gp = gpar(col = "black", lwd = 0.5),
      column_gap = unit(1, units = "mm"),
      row_names_gp = gpar(fontsize = 9),
      column_title_gp = gpar(fontsize = 12),
      row_names_side = "left",
      heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 10), title_gp = gpar(fontsize = 12),
        border = "lightgrey", title_position = "topcenter",
        grid_height = unit(4, "mm"), direction = "horizontal"
      )
    ),
    heatmap_legend_side = "top"
  )
  ht
}
```

</div>
<div>

```{r, fig.width = 7, fig.height = 8}
gse_plot <- "neurofibrillary tangle"
for (gse in gse_plot) {
  genes <- unlist(geneIds(gene_sets_sim[names(gene_sets_sim) %in% gse]))
  mat <- sim[rownames(sim) %in% genes, c(1, 6)]
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(mat), ]
  gene_anno_sub <- gene_anno_sub[match(rownames(mat), gene_anno_sub$ensembl), ]
  rownames(mat) <- gene_anno_sub$symbol
  counts(mat) <- t(
    apply(
      counts(mat), MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  ht <- draw(
    Heatmap(
      counts(mat),
      name = "Scaled Gene Expression",
      col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
      cluster_rows = FALSE,
      cluster_columns = FALSE,
      show_column_names = FALSE,
      top_annotation = top_anno4,
      column_split = mat$Group,
      column_title = gse,
      rect_gp = gpar(col = "black", lwd = 0.5),
      column_gap = unit(1, units = "mm"),
      row_names_gp = gpar(fontsize = 9),
      column_title_gp = gpar(fontsize = 12),
      row_names_side = "left",
      heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 10), title_gp = gpar(fontsize = 12),
        border = "lightgrey", title_position = "topcenter",
        grid_height = unit(4, "mm"), direction = "horizontal"
      )
    ),
    heatmap_legend_side = "top"
  )
  ht
}
```

</div>
</div>

### Low vs. High Variability

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r, fig.width = 7, fig.height = 8}
gse_plot <-
  "energy coupled proton transmembrane transport, against electrochemical gradient"
for (gse in gse_plot) {
  genes <- unlist(geneIds(gene_sets_sim[names(gene_sets_sim) %in% gse]))
  mat <- sim[rownames(sim) %in% genes, c(1, 2, 3)]
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(mat), ]
  gene_anno_sub <- gene_anno_sub[match(rownames(mat), gene_anno_sub$ensembl), ]
  rownames(mat) <- gene_anno_sub$symbol
  col_range <- c(
    min(counts(mat)),
    (min(counts(mat)) + max(counts(mat))) * 0.01,
    max(counts(mat))
  )
  ht <- draw(
    Heatmap(
      counts(mat),
      name = "Raw Counts",
      col = colorRamp2(col_range, colors = c("blue", "white", "red")),
      cluster_rows = FALSE,
      cluster_columns = FALSE,
      show_column_names = FALSE,
      top_annotation = top_anno5,
      column_split = mat$Group,
      column_title = gse,
      rect_gp = gpar(col = "black", lwd = 0.5),
      column_gap = unit(1, units = "mm"),
      row_names_gp = gpar(fontsize = 9),
      column_title_gp = gpar(fontsize = 12),
      row_names_side = "left",
      cell_fun = function(j, i, x, y, width, height, fill) {
        grid.text(
          sprintf("%.1f", counts(mat)[i, j]), x, y, gp = gpar(fontsize = 10)
        )
      },
      heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 10), title_gp = gpar(fontsize = 12),
        border = "lightgrey", title_position = "topcenter",
        grid_height = unit(4, "mm"), direction = "horizontal"
      )
    ),
    heatmap_legend_side = "top"
  )
  ht
}
```

</div>
<div>

```{r, fig.width = 7, fig.height = 8}
gse_plot <- "synaptic vesicle maturation"
for (gse in gse_plot) {
  genes <- unlist(geneIds(gene_sets_sim[names(gene_sets_sim) %in% gse]))
  mat <- sim[rownames(sim) %in% genes, c(1, 2, 3)]
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(mat), ]
  gene_anno_sub <- gene_anno_sub[match(rownames(mat), gene_anno_sub$ensembl), ]
  rownames(mat) <- gene_anno_sub$symbol
  col_range <- c(
    min(counts(mat)),
    (min(counts(mat)) + max(counts(mat))) * 0.001,
    (min(counts(mat)) + max(counts(mat))) * 0.925,
    max(counts(mat))
  )
  ht <- draw(
    Heatmap(
      counts(mat),
      name = "Raw Counts",
      col = colorRamp2(col_range, colors = c("blue", "lavender", "lightcoral", "red")),
      cluster_rows = FALSE,
      cluster_columns = FALSE,
      show_column_names = FALSE,
      top_annotation = top_anno5,
      column_split = mat$Group,
      column_title = gse,
      rect_gp = gpar(col = "black", lwd = 0.5),
      column_gap = unit(1, units = "mm"),
      row_names_gp = gpar(fontsize = 9),
      column_title_gp = gpar(fontsize = 12),
      row_names_side = "left",
      cell_fun = function(j, i, x, y, width, height, fill) {
        grid.text(
          sprintf("%.1f", counts(mat)[i, j]), x, y, gp = gpar(fontsize = 10)
        )
      },
      heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 10), title_gp = gpar(fontsize = 12),
        border = "lightgrey", title_position = "topcenter",
        grid_height = unit(4, "mm"), direction = "horizontal"
      )
    ),
    heatmap_legend_side = "top"
  )
  ht
}
```

</div>
</div>

### Differential Expression (NFT vs. CTRL)

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r, fig.width = 7, fig.height = 8}
gse_plot <- "tau protein binding"
for (gse in gse_plot) {
  genes <- unlist(geneIds(gene_sets_sim[names(gene_sets_sim) %in% gse]))
  mat <- sim[rownames(sim) %in% genes, c(7, 8)]
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(mat), ]
  gene_anno_sub <- gene_anno_sub[match(rownames(mat), gene_anno_sub$ensembl), ]
  rownames(mat) <- gene_anno_sub$symbol
  col_range <- c(
    min(logcounts(mat)),
    (min(logcounts(mat)) + max(logcounts(mat))) / 2,
    max(logcounts(mat))
  )
  ht <- draw(
    Heatmap(
      logcounts(mat),
      name = "log2 + 1 Gene Expression",
      col = colorRamp2(col_range, colors = c("blue", "white", "red")),
      cluster_rows = FALSE,
      cluster_columns = FALSE,
      show_column_names = FALSE,
      top_annotation = top_anno6,
      column_split = mat$Group,
      column_title = gse,
      rect_gp = gpar(col = "black", lwd = 0.5),
      column_gap = unit(1, units = "mm"),
      row_names_gp = gpar(fontsize = 9),
      column_title_gp = gpar(fontsize = 12),
      row_names_side = "left",
      heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 10), title_gp = gpar(fontsize = 12),
        border = "lightgrey", title_position = "topcenter",
        grid_height = unit(4, "mm"), direction = "horizontal"
      )
    ),
    heatmap_legend_side = "top"
  )
  ht
}
```

</div>
<div>

```{r, fig.width = 7, fig.height = 8}
gse_plot <- "CHOP-ATF4 complex"
for (gse in gse_plot) {
  genes <- unlist(geneIds(gene_sets_sim[names(gene_sets_sim) %in% gse]))
  mat <- sim[rownames(sim) %in% genes, c(7, 8)]
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(mat), ]
  gene_anno_sub <- gene_anno_sub[match(rownames(mat), gene_anno_sub$ensembl), ]
  rownames(mat) <- gene_anno_sub$symbol
  col_range <- c(
    min(logcounts(mat)),
    (min(logcounts(mat)) + max(logcounts(mat))) / 2,
    max(logcounts(mat))
  )
  ht <- draw(
    Heatmap(
      logcounts(mat),
      name = "log2 + 1 Gene Expression",
      col = colorRamp2(col_range, colors = c("blue", "white", "red")),
      cluster_rows = FALSE,
      cluster_columns = FALSE,
      show_column_names = FALSE,
      top_annotation = top_anno6,
      column_split = mat$Group,
      column_title = gse,
      rect_gp = gpar(col = "black", lwd = 0.5),
      column_gap = unit(1, units = "mm"),
      row_names_gp = gpar(fontsize = 9),
      column_title_gp = gpar(fontsize = 12),
      row_names_side = "left",
      heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 10), title_gp = gpar(fontsize = 12),
        border = "lightgrey", title_position = "topcenter",
        grid_height = unit(4, "mm"), direction = "horizontal"
      )
    ),
    heatmap_legend_side = "top"
  )
  ht
}
```

</div>
</div>

Performance Benchmarks
===

Row {.tabset}
---

### Real Dataset (Pseudobulk, 6 Samples, Serial Processing)

```{r}
bench[ , 1] <- names(out_sim)
bench[ , 1:9]
```

### Real Dataset (Pseudobulk, 60 Samples, Parallel Processing)

```{r}
bench_big[ , 1] <- names(out_sim)
bench_big[ , 1:9]
```

### Real Dataset (Single-cell, 600 Samples, Parallel Processing)

```{r}
bench_sc[ , 1] <- names(out_sim)
bench_sc[ , 1:9]

bench_sc2[ , 1] <- names(out_sim)
bench_sc2[ , 1:9]

bench_sc3[ , 1] <- names(out_sim)
bench_sc3[ , 1:9]

bench_sc4[ , 1] <- names(out_sim)
bench_sc4[ , 1:9]

bench_sc5[ , 1] <- names(out_sim)
bench_sc5[ , 1:9]

bench_sc6[ , 1] <- names(out_sim)
bench_sc6[ , 1:9]

bench_comb <- list(
  bench_sc, bench_sc2, bench_sc3, bench_sc4, bench_sc5, bench_sc6
)
bench_comb <- do.call(
  rbind,
  lapply(
    seq_along(bench_comb),
    FUN = function(i) {
      cbind(bench_comb[[i]], subset = paste0("subset", i))
    }
  )
)
bench_comb$subset <- factor(
  bench_comb$subset, levels = unique(bench_comb$subset)
)

time_formatter <- function(x) {
  minutes <- floor(x / 60)
  seconds <- x %% 60
  paste0(minutes, "m ", round(seconds), "s")
}
breaks <- c(
  3, 4, 5, 6, 7, 10, 20, 30, 50, 70, 100, 200, 300, 500, 700, 1000, 1500, 3000
)

ggplot(
  bench_comb, aes(
    x = subset, y = median, color = expression, group = expression
  )
) +
  geom_point(size = 3) +
  geom_line() +
  scale_y_log10(breaks = breaks, labels = time_formatter) +
  labs(
    title = "Benchmarking Tools on Different Data Subsets",
    x = "Data Subsets",
    y = "Time Elapsed",
    color = "Expression"
  ) +
  theme_minimal()
```

Row {.tabset}
---

```{r, include = FALSE}
rm(
  ht_list, loc1, loc2, mat, out, row_anno, tmp, top_anno, top_anno2, top_anno3,
  top_anno4, top_anno5, top_anno6, top_anno7, top_anno8, out_sim, gene_sets_sim,
  gene_anno, ht, dge_orig, dge, gene_anno_sub, gene_sets, grid_list, sim,
  default_mat, sc, out_sc, bench, bench_big, bench_sc, bench_sc_big, sim_sub,
  sim_big, sc_big, bench_sc2, bench_sc3, bench_sc4, bench_sc5, bench_comb
)
gc()
```
