---
title: "GeneFunnel Benchmarks Viewer"
output:
  flexdashboard::flex_dashboard:
    source_code: embed
    orientation: rows
    runtime: shiny
---

```{r global, include = FALSE}
#    This file is part of genefunnel-benchmarks.
#    Copyright (C) 2024  Emir Turkes, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

library(conflicted)
packages <- c(
  "Rcpp", "Matrix", "shinyMatrix", "ComplexHeatmap", "circlize", "GSVA",
  "BiocParallel", "parallelly", "cowplot", "knitr", "kableExtra"
)
lapply(packages, FUN = library, character.only = TRUE)

sourceCpp(file.path("..", "src", "calculateScores.cpp"))
source("utils.R")

top_anno <- HeatmapAnnotation(
  sample = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = c("Sample 1", "Sample 2", "Sample 3", "Sample 4", "Sample 5"),
    labels_gp = gpar(col = "black", fontsize = 15),
    height = unit(10, "mm")
  )
)
top_anno2 <- HeatmapAnnotation(
  sample = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = c("Sample 1", "Sample 2", "Sample 3", "Sample 4"),
    labels_gp = gpar(col = "black", fontsize = 15),
    height = unit(10, "mm")
  )
)
top_anno3 <- HeatmapAnnotation(
  sample = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = c("S1", "S2", "S3", "S4"),
    labels_gp = gpar(col = "black", fontsize = 15),
    height = unit(10, "mm")
  )
)
row_anno <- rowAnnotation(
  gene_sets = anno_block(labels = c("Gene Set X", "Gene Set Y")),
  empty = anno_empty(border = FALSE, width = unit(8, units = "mm"))
)

colnames <- c(
  "------Sample 1",
  "------Sample 2",
  "------Sample 3",
  "------Sample 4"
)

vec <- c(
  rep(50, times = 10),
  c(90, 10, 90, 10, 90, 90, 10, 10, 90, 10),
  rep(100, times = 10),
  rep(100, times = 5), rep(0, times = 5),
  rep(100, times = 5), rep(NA, times = 5)
)
set.seed(4)
noise1 <- round(runif(10, min = -5, max = 5))
noise1[1] <- 2
noise2 <- noise1 * 2
noise2[1] <- -90
noise2[3] <- noise2[3] + 42
noise2[5] <- noise2[5] + 42
noise2[7] <- noise2[7] + 10
noise2 <- noise2[c(1:5, 7, 6, 8:10)]
noise3 <- noise1 * 2
noise4 <- rep(0, times = 10)
noise5 <- noise4

default_mat <- matrix(
  vec + c(noise1, noise2, noise3, noise4, noise5),
  nrow =  10,
  ncol = 5,
  dimnames = list(
    paste("Gene", toupper(letters[1:10])),
    c("Sample1", "Sample2", "Sample3", "Sample4", "Sample5")
  )
)
gene_sets <- list(
  `Gene Set X` = rownames(default_mat)[1:5],
  `Gene Set Y` = rownames(default_mat)[6:10],
  `Gene Set Z` = rownames(default_mat)
)

ht_opt$DIMNAME_PADDING <- unit(3, units = "mm")
```

Algorithm Overview
===

Row {data-height=355}
---

### About GeneFunnel 1 {.no-title}

GeneFunnel is a tool for gene set enrichment (or feature set scoring to generalise the concept), falling in the class of methods commonly referred to as ssGSEA-like (short for single-sample gene set enrichment analysis).
It takes as input a matrix of genes/proteins (generally referred to as features) and samples, for example the output of an (sc)RNAseq or mass-spectrometry proteomics experiment.
Also required is a gene set list, such as those provided by the Gene Ontology (GO), where each element is a set of genes (or any feature matching those in the input matrix) with a name for each set.
The algorithm for GeneFunnel is expressed mathematically below, and explained intuitively in the right-hand pane.

$$
\text{score}_{k,j} = \sum_{i \in G_k} X_{i,j} - \sum_{i \in G_k} \left| X_{i,j} - \bar{X}_{G_k,j} \right|
$$
Where:

- $\sum_{i \in G_k} X_{i,j}$ is the sum of the expression levels for the genes in gene set $G_k$ for sample $j$.
- $\bar{X}_{G_k,j}$ is the mean expression of the genes in gene set $G_k$ for sample $j$.
- $\sum_{i \in G_k} \left| X_{i,j} - \bar{X}_{G_k,j} \right|$ is the sum of the absolute deviations from the mean.

### About GeneFunnel 2 {.no-title}

GeneFunnel iterates through samples in the matrix, where for each sample, the feature list is iterated through.
For each set in the list, the matrix is subset to the matching features and scored.
Conceptually, scoring takes the sum of the matching features, then adjusts the sum based on the deviance of each feature from the mean of the matching features.
Specifically, the measure of deviance is taken by subtracting the value of each feature from the mean of matched features, taking the absolute value, then summing all of the deviance values.
The absolute value is neccessary as many features will have values lower than the mean, and we are interested in any deviance whether positive or negative.
  
The summed deviance is then substracted from the overall sum, providing the final score.
This provides the property that a minimally deviant set will be equal to the sum, while a maximally deviant set will be equal to the inverse sum, in other words the sum multipled by -1.
Users may be interested excluding sets with negative values or clamping their scores to zero, as such sets are considered unenriched.
It is also up to the discretion of the user to log or otherwise transform scores, as is done with scRNAseq data to compress values to a smaller range, though in most cases the deviance correction will naturally protect against extreme values.

Row {data-height=650}
---

### Interactive Example {.no-title}

> Below is a contrived scenario of genes and samples with the resulting GeneFunnel scoring on the right and associated metrics.
> Try editing some values to see it update in real-time!

```{r}
shinyApp(
  ui = fluidPage(
    mainPanel(
      width = 3,
      tags$h3(
        HTML("тож Click Cells to Edit Matrix тоз"), style = "text-align:center"
      ),
      matrixInput(
        "mat",
        label = "Hypothetical Raw Gene Count X Sample Matrix",
        value = default_mat
      )
    ),
    mainPanel(
      width = 4,
      plotOutput("gene_heatmap", height = "400px")
    ),
    mainPanel(
      width = 5,
      plotOutput("gse_heatmap", height = "400px")
    )
  ),
  server = function(input, output, session) {

    output$gene_heatmap <- renderPlot({
      heatmap_mat <- apply(input$mat, MARGIN = 2, FUN = as.numeric)
      rownames(heatmap_mat) <- rownames(input$mat)

      col_range <- c(
        min(heatmap_mat, na.rm = TRUE),
        (min(heatmap_mat, na.rm = TRUE) + max(heatmap_mat, na.rm = TRUE)) / 2,
        max(heatmap_mat, na.rm = TRUE)
      )

      draw(
        Heatmap(
          heatmap_mat,
          name = "Raw Gene Counts",
          col = colorRamp2(col_range, colors = c("blue", "white", "red")),
          cluster_rows = FALSE,
          cluster_columns = FALSE,
          show_column_names = FALSE,
          top_annotation = top_anno,
          column_split = seq(5),
          column_title = "Hypothetical Raw Gene Count by Sample Matrix",
          rect_gp = gpar(col = "black", lwd = 0.5),
          column_gap = unit(1, units = "mm"),
          row_names_gp = gpar(fontsize = 15),
          column_title_gp = gpar(fontsize = 18),
          row_names_side = "left",
          heatmap_legend_param = list(
            labels_gp = gpar(fontsize = 14), title_gp = gpar(fontsize = 15),
            border = "lightgrey", title_position = "topcenter",
            grid_height = unit(4, units = "mm"), direction = "horizontal"
          )
        ),
        heatmap_legend_side = "top"
      )
    })

    output$gse_heatmap <- renderPlot({
      numeric_mat <- apply(input$mat, MARGIN = 2, FUN = as.numeric)
      heatmap_mat <- matrix(
        c(
          sum(numeric_mat[ , 1], na.rm = TRUE) -
            sum(
              abs(numeric_mat[ , 1] - mean(numeric_mat[ , 1], na.rm = TRUE)),
              na.rm = TRUE
            ),
          sum(numeric_mat[ , 2], na.rm = TRUE) -
            sum(
              abs(numeric_mat[ , 2] - mean(numeric_mat[ , 2], na.rm = TRUE)),
              na.rm = TRUE
            ),
          sum(numeric_mat[ , 3], na.rm = TRUE) -
            sum(
              abs(numeric_mat[ , 3] - mean(numeric_mat[ , 3], na.rm = TRUE)),
              na.rm = TRUE
            ),
          sum(numeric_mat[ , 4], na.rm = TRUE) -
            sum(
              abs(numeric_mat[ , 4] - mean(numeric_mat[ , 4], na.rm = TRUE)),
              na.rm = TRUE
            ),
          sum(numeric_mat[ , 5], na.rm = TRUE) -
            sum(
              abs(numeric_mat[ , 5] - mean(numeric_mat[ , 5], na.rm = TRUE)),
              na.rm = TRUE
            )
        ),
        nrow =  1,
        ncol = 5,
        dimnames = list("Gene Set X  \n(Genes A to J)")
      )

      bottom_anno <- HeatmapAnnotation(
        text = anno_text(
          c(
            paste0(
              "\n\n\n\nSum = ", sum(numeric_mat[ , 1], na.rm = TRUE),
              "\nMean = ", mean(numeric_mat[ , 1], na.rm = TRUE),
              "\nVar = ", round(
                sum(
                  abs(
                    numeric_mat[ , 1] - mean(numeric_mat[ , 1], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                ),
                digits = 1
              ),
              "\nScore = ", sum(numeric_mat[ , 1], na.rm = TRUE) -
                sum(
                  abs(
                    numeric_mat[ , 1] - mean(numeric_mat[ , 1], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                )
            ),
            paste0(
              "\n\n\n\nSum = ", sum(numeric_mat[ , 2], na.rm = TRUE),
              "\nMean = ", mean(numeric_mat[ , 2], na.rm = TRUE),
              "\nVar = ", round(
                sum(
                  abs(
                    numeric_mat[ , 2] - mean(numeric_mat[ , 2], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                ),
                digits = 1
              ),
              "\nScore = ", sum(numeric_mat[ , 2], na.rm = TRUE) -
                sum(
                  abs(
                    numeric_mat[ , 2] - mean(numeric_mat[ , 2], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                )
            ),
            paste0(
              "\n\n\n\nSum = ", sum(numeric_mat[ , 3], na.rm = TRUE),
              "\nMean = ", mean(numeric_mat[ , 3], na.rm = TRUE),
               "\nVar = ", round(
                sum(
                  abs(
                    numeric_mat[ , 3] - mean(numeric_mat[ , 3], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                ),
                digits = 1
              ),
              "\nScore = ", sum(numeric_mat[ , 3], na.rm = TRUE) -
                sum(
                  abs(
                    numeric_mat[ , 3] - mean(numeric_mat[ , 3], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                )
            ),
            paste0(
              "\n\n\n\nSum = ", sum(numeric_mat[ , 4], na.rm = TRUE),
              "\nMean = ", mean(numeric_mat[ , 4], na.rm = TRUE),
              "\nVar = ", round(
                sum(
                  abs(
                    numeric_mat[ , 4] - mean(numeric_mat[ , 4], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                ),
                digits = 1
              ),
              "\nScore = ", sum(numeric_mat[ , 4], na.rm = TRUE) -
                sum(
                  abs(
                    numeric_mat[ , 4] - mean(numeric_mat[ , 4], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                )
            ),
            paste0(
              "\n\n\n\nSum = ", sum(numeric_mat[ , 5], na.rm = TRUE),
              "\nMean = ", mean(numeric_mat[ , 5], na.rm = TRUE),
              "\nVar = ", round(
                sum(
                  abs(
                    numeric_mat[ , 5] - mean(numeric_mat[ , 5], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                ),
                digits = 1
              ),
              "\nScore = ", sum(numeric_mat[ , 5], na.rm = TRUE) -
                sum(
                  abs(
                    numeric_mat[ , 5] - mean(numeric_mat[ , 5], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                )
            )
          ),
          rot = 0,
          gp = gpar(fontsize = 18),
          just = "centre"
        )
      )

      col_range <- c(
        min(heatmap_mat, na.rm = TRUE),
        (min(heatmap_mat, na.rm = TRUE) + max(heatmap_mat, na.rm = TRUE)) / 2,
        max(heatmap_mat, na.rm = TRUE)
      )

      draw(
        Heatmap(
          heatmap_mat,
          name = "Gene Set Score",
          col = colorRamp2(col_range, colors = c("blue", "white", "red")),
          cluster_rows = FALSE,
          cluster_columns = FALSE,
          show_column_names = FALSE,
          top_annotation = top_anno,
          bottom_annotation = bottom_anno,
          column_split = seq(5),
          column_title = "Resulting Gene Set Enrichment by Sample Matrix",
          rect_gp = gpar(col = "black", lwd = 0.5),
          column_gap = unit(1, units = "mm"),
          row_names_gp = gpar(fontsize = 16),
          column_title_gp = gpar(fontsize = 18),
          row_names_side = "left",
          height = unit(0.5, "npc"),
          heatmap_legend_param = list(
            labels_gp = gpar(fontsize = 14), title_gp = gpar(fontsize = 15),
            border = "lightgrey", title_position = "topcenter",
            grid_height = unit(4, units = "mm"), direction = "horizontal"
          )
        ),
        heatmap_legend_side = "top",
        padding = unit(c(75, 2, 2, 2), units = "mm")
      )
    })
  }
)
```

Comparison with Other Methods
===

Row {data-height=500}
---

### Example Matrix {.no-title}

```{r}
mat <- default_mat[ , 1:4]

col_range <- c(min(mat), (min(mat) + max(mat)) / 2, max(mat))
draw(
  Heatmap(
    mat,
    name = "Raw Gene Counts",
    col = colorRamp2(col_range, colors = c("blue", "white", "red")),
    cluster_rows = FALSE,
    cluster_columns = FALSE,
    show_column_names = FALSE,
    row_title = NULL,
    top_annotation = top_anno2,
    right_annotation = row_anno,
    column_split = seq(4),
    row_split = c(rep(1, times = 5), rep(2, times = 5)),
    column_title = "Hypothetical Raw Gene Count by Sample Matrix",
    rect_gp = gpar(col = "black", lwd = 0.5),
    column_gap = unit(1, units = "mm"),
    row_names_gp = gpar(fontsize = 15),
    column_title_gp = gpar(fontsize = 16),
    row_names_side = "left",
    heatmap_legend_param = list(
      labels_gp = gpar(fontsize = 14), title_gp = gpar(fontsize = 15),
      border = "lightgrey", title_position = "topcenter",
      grid_height = unit(4, units = "mm"), direction = "horizontal"
    )
  ),
  heatmap_legend_side = "top"
)

seekViewport("annotation_empty_1")
loc1 = deviceLoc(x = unit(0, "npc"), y = unit(0, "npc"))
seekViewport("annotation_empty_2")
loc2 = deviceLoc(x = unit(1, "npc"), y = unit(1, "npc"))

seekViewport("global")
grid.rect(
  loc2$x * 0.95, (loc1$y + loc2$y) / 2,
  width = loc2$x - loc1$x, height = loc2$y * 1.93,
  just = c("left", "centre")
)
grid.text(
  "Gene Set Z", rot = 90,
  x = (loc1$x + loc2$x) * 0.501, y = (loc1$y + loc2$y) * 0.5
)
```

### Background {.no-title}

> On this page, the hypothetical gene count matrix shown left is compared across 5 different single-sample GSEA methods and GeneFunnel.  
>  
> The gene count matrix is the same as the one in Algorithm Overview except the last sample is excluded because most of the methods cannot handle NaN values.  
>  
> Three types of gene sets are considered, as annotated on the right-hand side of the count matrix: the first 5 genes, the last 5 genes, and all 10 genes.  
>  
> Heatmaps for each method is shown in the bottom-left and to the right are the raw numerical output tables.

```{r}
```

Row {data-height=555}
---

### Gene Set Scoring {.no-title}

```{r, results = "hide"}
out <- vector("list", length = 6)
names(out) <- c(
  "GSVA Poisson", "GSVA Gaussian", "ssGSEA", "PLAGE", "Z-score", "GeneFunnel"
)

tmp <- gsvaParam(mat, gene_sets, kcdf = "Poisson")
out[[1]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

tmp <- gsvaParam(log2(mat + 1), gene_sets, kcdf = "Gaussian")
out[[2]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

tmp <- ssgseaParam(log2(mat + 1), gene_sets, normalize = FALSE)
out[[3]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

tmp <- zscoreParam(log2(mat + 1), gene_sets)
out[[4]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

tmp <- plageParam(log2(mat + 1), gene_sets)
out[[5]] <- gsva(tmp, BPPARAM = MulticoreParam(availableCores()))

out[[6]] <- genefunnel(
  mat, gene_sets, BPPARAM = MulticoreParam(availableCores())
)
```

```{r, fig.width = 7}
grid_list <- vector("list", length = 2)

ht_list = NULL
for (i in c(1:3)) {
  col_range <- c(
    min(out[[i]], na.rm = TRUE),
    (min(out[[i]], na.rm = TRUE) + max(out[[i]], na.rm = TRUE)) / 2,
    max(out[[i]], na.rm = TRUE)
  )
  ht_list <- ht_list +
    Heatmap(
      out[[i]],
      col = colorRamp2(col_range, colors = c("blue", "white", "red")),
      name = names(out)[[i]],
      cluster_rows = FALSE,
      cluster_columns = FALSE,
      show_column_names = FALSE,
      top_annotation = top_anno3,
      column_split = seq(4),
      column_title = NULL,
      rect_gp = gpar(col = "black", lwd = 0.5),
      column_gap = unit(1, units = "mm"),
      row_names_gp = gpar(fontsize = 16),
      column_title_gp = gpar(fontsize = 18),
      row_names_side = "left",
      width = unit(4, units = "cm"),
      height = unit(2, units = "cm"),
      heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 14), title_gp = gpar(fontsize = 15),
        border = "lightgrey", title_position = "topcenter",
        grid_height = unit(4, units = "mm"), direction = "horizontal",
        legend_width = unit(3.75, units = "cm")
      )
    )
}
grid_list[[1]] <- grid.grabExpr(
  draw(
    ht_list, heatmap_legend_side = "top", legend_gap = unit(0.75, units = "cm")
  )
)

ht_list = NULL
for (i in c(4:6)) {
  col_range <- c(
    min(out[[i]], na.rm = TRUE),
    (min(out[[i]], na.rm = TRUE) + max(out[[i]], na.rm = TRUE)) / 2,
    max(out[[i]], na.rm = TRUE)
  )
  ht_list <- ht_list +
    Heatmap(
      out[[i]],
      col = colorRamp2(col_range, colors = c("blue", "white", "red")),
      name = names(out)[[i]],
      cluster_rows = FALSE,
      cluster_columns = FALSE,
      show_column_names = FALSE,
      top_annotation = top_anno3,
      column_split = seq(4),
      column_title = NULL,
      rect_gp = gpar(col = "black", lwd = 0.5),
      column_gap = unit(1, units = "mm"),
      row_names_gp = gpar(fontsize = 16),
      column_title_gp = gpar(fontsize = 18),
      row_names_side = "left",
      width = unit(4, units = "cm"),
      height = unit(2, units = "cm"),
      heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 14), title_gp = gpar(fontsize = 15),
        border = "lightgrey", title_position = "topcenter",
        grid_height = unit(4, units = "mm"), direction = "horizontal",
        legend_width = unit(3.75, units = "cm")
      )
    )
}
grid_list[[2]] <- grid.grabExpr(
  draw(
    ht_list, heatmap_legend_side = "top", legend_gap = unit(0.75, units = "cm")
  )
)

plot_grid(grid_list[[1]], grid_list[[2]], nrow = 2, vjust = 1)
```

### Scoring Output 1 {.no-title}

```{r, results = "asis"}
for (i in c(1:3)) {
  print(
    kable_styling(
      kable(out[[i]], col.names = colnames, caption = names(out)[i]),
      bootstrap_options = "none"
    )
  )
}
```

### Scoring Output 2 {.no-title}

```{r, results = "asis"}
for (i in c(4:6)) {
  print(
    kable_styling(
      kable(out[[i]], col.names = colnames, caption = names(out)[i]),
      bootstrap_options = "none"
    )
  )
}
```

```{r, include = FALSE}
rm(ht_list, loc1, loc2, mat, out, row_anno, tmp, top_anno2, top_anno3)
gc()
```
