---
title: "GeneFunnel Benchmarks Viewer"
output:
  flexdashboard::flex_dashboard:
    source_code: embed
    orientation: rows
    runtime: shiny
---

```{r global, include = FALSE}
#    This file is part of genefunnel-benchmarks.
#    Copyright (C) 2024  Emir Turkes, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

library(conflicted)
packages <- c(
  "shinyMatrix", "ComplexHeatmap", "circlize"
)
lapply(packages, FUN = library, character.only = TRUE)

top_anno <- HeatmapAnnotation(
  sample = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = c("Sample 1", "Sample 2", "Sample 3", "Sample 4"),
    labels_gp = gpar(col = "black", fontsize = 15),
    height = unit(10, "mm")
  )
)

vec <- c(
  rep(50, times = 10), rep(c(97, 3), times = 5),
  rep(100, times = 10), rep(101, times = 5), rep(NA, times = 5)
)
set.seed(1)
noise1 <- round(runif(20, min = -4, max = 4))
noise2 <- noise1[1:10] * 2
set.seed(1)
noise3 <- round(runif(10, min = -4, max = 4))

default_mat <- matrix(
  vec + c(noise1, noise2, noise3),
  nrow =  10,
  ncol = 4,
  dimnames = list(
    paste("Gene", toupper(letters[1:10])),
    c("Sample 1", "Sample 2", "Sample 3", "Sample 4")
  )
)

ht_opt$DIMNAME_PADDING = unit(3, units = "mm")
```

Algorithm Overview
===

Row {data-height=400}
---

### About GeneFunnel

- GeneFunnel is a tool for gene set enrichment (or feature set scoring to generalise the concept), falling in the class of methods known as single-sample GSEA (ssGSEA).  
- It takes a matrix of genes/proteins (generally referred to as features) and samples, for example the output of an (sc)RNAseq or mass-spectrometry proteomics experiment.  
- A list, where each element is a set of genes (or any feature matching those in the input matrix) with a name for the set, is also provided.  
- A common example of such kind of list are those provided by the Gene Ontology (GO), whose aim is to group together genes into functionally related sets.  
- The algorithm iterates through samples in the matrix, where for each sample, the feature list is iterated through.  
- For each set in the list, the matrix is subset to the matching features and scored.  
- Conceptually, scoring takes the sum of the matching features, then adjusts the sum based on the deviance of each feature from the mean of the matching features.  
- Specifically, the measure of deviance is taken by subtracting the value of each feature from the mean of matched features, taking the absolute value, then summing all of the deviance values.  
- The absolute value is neccessary as many features will have values lower than the mean, and we are interested in any deviance whether positive or negative.  
- The summed deviance is then substracted from the overall sum, however, the overall sum is first multipled by 2.  
- This is to ensure that the final score is a positive value and provides the property that a maximally deviant set will simply produce a score equal to the sum of features, while a minimally deviant set will produce a score equal to twice the sum of features.  
- This algorithm expressed mathematically is provided below:  

$$
\text{score}_{k,j} = 2 \times \sum_{i \in G_k} X_{i,j} - \sum_{i \in G_k} \left| X_{i,j} - \bar{X}_{G_k,j} \right|
$$

Row {data-height=600}
---

### Interactive Example

> A contrived scenario of genes and samples with the resulting GeneFunnel scoring on the right and associated metrics.
> Try editing some values to see it update in real-time!

```{r}
shinyApp(
  ui = fluidPage(
    sidebarPanel(
      width = 3,
      tags$h3(
        HTML("тож Click Cells to Edit Matrix тоз"), style = "text-align:center"
      ),
      matrixInput(
        "mat",
        label = "Hypothetical Raw Gene Count X Sample Matrix",
        value = default_mat
      )
    ),
    mainPanel(
      width = 4,
      plotOutput("gene_heatmap", height = "450px")
    ),
    mainPanel(
      width = 5,
      plotOutput("gse_heatmap", height = "450px")
    )
  ),
  server = function(input, output, session) {

    output$gene_heatmap <- renderPlot({
      heatmap_mat <- apply(input$mat, MARGIN = 2, FUN = as.numeric)
      rownames(heatmap_mat) <- rownames(input$mat)

      col_range <- c(
        min(heatmap_mat, na.rm = TRUE),
        (min(heatmap_mat, na.rm = TRUE) + max(heatmap_mat, na.rm = TRUE)) / 2,
        max(heatmap_mat, na.rm = TRUE)
      )

      draw(
        Heatmap(
          heatmap_mat,
          name = "Gene Expression",
          col = colorRamp2(col_range, colors = c("blue", "white", "red")),
          cluster_rows = FALSE,
          cluster_columns = FALSE,
          show_column_names = FALSE,
          top_annotation = top_anno,
          column_split = seq(4),
          column_title = "Hypothetical Raw Gene Count by Sample Matrix",
          rect_gp = gpar(col = "black", lwd = 0.5),
          column_gap = unit(1, units = "mm"),
          row_names_gp = gpar(fontsize = 15),
          column_title_gp = gpar(fontsize = 18),
          row_names_side = "left",
          heatmap_legend_param = list(
            labels_gp = gpar(fontsize = 14), title_gp = gpar(fontsize = 15),
            border = "lightgrey", title_position = "topcenter",
            grid_height = unit(4, "mm"), direction = "horizontal"
          )
        ),
        heatmap_legend_side = "top"
      )
    })

    output$gse_heatmap <- renderPlot({
      numeric_mat <- apply(input$mat, MARGIN = 2, FUN = as.numeric)
      heatmap_mat <- matrix(
        c(
          sum(numeric_mat[ , 1], na.rm = TRUE) * 2 -
            sum(
              abs(numeric_mat[ , 1] - mean(numeric_mat[ , 1], na.rm = TRUE)),
              na.rm = TRUE
            ),
          sum(numeric_mat[ , 2], na.rm = TRUE) * 2 -
            sum(
              abs(numeric_mat[ , 2] - mean(numeric_mat[ , 2], na.rm = TRUE)),
              na.rm = TRUE
            ),
          sum(numeric_mat[ , 3], na.rm = TRUE) * 2 -
            sum(
              abs(numeric_mat[ , 3] - mean(numeric_mat[ , 3], na.rm = TRUE)),
              na.rm = TRUE
            ),
          sum(numeric_mat[ , 4], na.rm = TRUE) * 2 -
            sum(
              abs(numeric_mat[ , 4] - mean(numeric_mat[ , 4], na.rm = TRUE)),
              na.rm = TRUE
            )
        ),
        nrow =  1,
        ncol = 4,
        dimnames = list("Gene Set X  \n(Genes A to J)")
      )

      bottom_anno <- HeatmapAnnotation(
        text = anno_text(
          c(
            paste0(
              "\n\n\n\nSum = ", sum(numeric_mat[ , 1], na.rm = TRUE) * 2,
              "\nMean = ", mean(numeric_mat[ , 1], na.rm = TRUE),
              "\nVar = ", round(
                sum(
                  abs(
                    numeric_mat[ , 1] - mean(numeric_mat[ , 1], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                ),
                digits = 1
              ),
              "\nScore = ", sum(numeric_mat[ , 1], na.rm = TRUE) * 2 -
                sum(
                  abs(
                    numeric_mat[ , 1] - mean(numeric_mat[ , 1], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                )
            ),
            paste0(
              "\n\n\n\nSum = ", sum(numeric_mat[ , 2], na.rm = TRUE) * 2,
              "\nMean = ", mean(numeric_mat[ , 2], na.rm = TRUE),
              "\nVar = ", round(
                sum(
                  abs(
                    numeric_mat[ , 2] - mean(numeric_mat[ , 2], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                ),
                digits = 1
              ),
              "\nScore = ", sum(numeric_mat[ , 2], na.rm = TRUE) * 2 -
                sum(
                  abs(
                    numeric_mat[ , 2] - mean(numeric_mat[ , 2], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                )
            ),
            paste0(
              "\n\n\n\nSum = ", sum(numeric_mat[ , 3], na.rm = TRUE) * 2,
              "\nMean = ", mean(numeric_mat[ , 3], na.rm = TRUE),
               "\nVar = ", round(
                sum(
                  abs(
                    numeric_mat[ , 3] - mean(numeric_mat[ , 3], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                ),
                digits = 1
              ),
              "\nScore = ", sum(numeric_mat[ , 3], na.rm = TRUE) * 2 -
                sum(
                  abs(
                    numeric_mat[ , 3] - mean(numeric_mat[ , 3], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                )
            ),
            paste0(
              "\n\n\n\nSum = ", sum(numeric_mat[ , 4], na.rm = TRUE) * 2,
              "\nMean = ", mean(numeric_mat[ , 4], na.rm = TRUE),
              "\nVar = ", round(
                sum(
                  abs(
                    numeric_mat[ , 4] - mean(numeric_mat[ , 4], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                ),
                digits = 1
              ),
              "\nScore = ", sum(numeric_mat[ , 4], na.rm = TRUE) * 2 -
                sum(
                  abs(
                    numeric_mat[ , 4] - mean(numeric_mat[ , 4], na.rm = TRUE)
                  ),
                  na.rm = TRUE
                )
            )
          ),
          rot = 0,
          gp = gpar(fontsize = 18),
          just = "centre"
        )
      )

      col_range <- c(
        min(heatmap_mat, na.rm = TRUE),
        (min(heatmap_mat, na.rm = TRUE) + max(heatmap_mat, na.rm = TRUE)) / 2,
        max(heatmap_mat, na.rm = TRUE)
      )

      draw(
        Heatmap(
          heatmap_mat,
          name = "Gene Set Score",
          col = colorRamp2(col_range, colors = c("blue", "white", "red")),
          cluster_rows = FALSE,
          cluster_columns = FALSE,
          show_column_names = FALSE,
          top_annotation = top_anno,
          bottom_annotation = bottom_anno,
          column_split = seq(4),
          column_title = "Resulting Gene Set Enrichment by Sample Matrix",
          rect_gp = gpar(col = "black", lwd = 0.5),
          column_gap = unit(1, units = "mm"),
          row_names_gp = gpar(fontsize = 16),
          column_title_gp = gpar(fontsize = 18),
          row_names_side = "left",
          height = unit(0.5, "npc"),
          heatmap_legend_param = list(
            labels_gp = gpar(fontsize = 14), title_gp = gpar(fontsize = 15),
            border = "lightgrey", title_position = "topcenter",
            grid_height = unit(4, "mm"), direction = "horizontal"
          )
        ),
        heatmap_legend_side = "top",
        padding = unit(c(90, 2, 2, 2), units = "mm")
      )
    })
  }
)
```

```{r, include = FALSE}
rm(vec, noise)
gc()
```
